{"cells":[{"cell_type":"markdown","id":"52d949fd","metadata":{"id":"52d949fd"},"source":["# Python-based pipeline for preprocessing of RNA-seq datasets"]},{"cell_type":"markdown","id":"b513920b","metadata":{"id":"b513920b"},"source":["## Protocol B-1 - Data Pre-processing for RNA-seq dataset"]},{"cell_type":"markdown","id":"dfe96e96","metadata":{"id":"dfe96e96"},"source":["## Step 1. Set-up\n","\n","### Install and import libraries\n","\n","Install the packages using `pip`.\n","If you want to install the library, then type  \"pip3 install `package name`\" in terminal."]},{"cell_type":"code","execution_count":null,"id":"da640717","metadata":{"id":"da640717","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728324707067,"user_tz":-540,"elapsed":68660,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"8fa959bf-dac8-46f2-e153-bb359b817f74"},"outputs":[{"output_type":"stream","name":"stdout","text":["numpy is installed.\n","Bio is not installed. Installing biopython...\n","pysam is not installed. Installing pysam...\n","pandas is installed.\n","samtools is not installed. Installing samtools...\n","bedtools is not installed. Installing bedtools...\n","bowtie2 is not installed. Installing bowtie2...\n"]}],"source":["import os, math, shutil, time, datetime, importlib\n","\n","libraries = {\n","    'numpy': 'numpy',\n","    'Bio': 'biopython',\n","    'pysam': 'pysam',\n","    'pandas': 'pandas'\n","}\n","\n","# Check and install Python libraries\n","for lib, package_name in libraries.items():\n","    try:\n","        importlib.import_module(lib)\n","        print(f\"{lib} is installed.\")\n","    except ImportError:\n","        print(f\"{lib} is not installed. Installing {package_name}...\")\n","        os.system(f'pip3 install {package_name}')\n","\n","# Binary tools to check and install\n","binary_tools = ['samtools', 'bedtools', 'bowtie2']\n","\n","# Check and install binary tools\n","for tool_name in binary_tools:\n","    if os.system(f'which {tool_name} > /dev/null 2>&1') == 0:\n","        print(f\"{tool_name} is already installed.\")\n","    else:\n","        print(f\"{tool_name} is not installed. Installing {tool_name}...\")\n","        os.system(f'apt-get update && apt-get install -y {tool_name}')"]},{"cell_type":"code","source":["from numpy import zeros\n","from Bio import SeqIO\n","import pandas as pd\n","import pysam"],"metadata":{"id":"vfDd8i68NRWi"},"id":"vfDd8i68NRWi","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"a434b0e8","metadata":{"id":"a434b0e8"},"source":["### Function\n","\n","Please run the cell without modification"]},{"cell_type":"code","execution_count":null,"id":"8c156b44","metadata":{"id":"8c156b44"},"outputs":[],"source":["def count_coverage(samfile, chromosome_size=6000000, flip=True):\n","    \"\"\"counts coverage per base in a strand-specific manner\n","\n","    For paired-end reads, the insert betwen the mapped reads is\n","    also counted.\n","\n","    flip: Whether or not the strands should be flipped.\n","    This should be true for RNA-seq, and false for ChIP-exo\n","\n","    chromosome_size: This value should be larger than the largest chromosome\"\"\"\n","\n","    if samfile.has_index():\n","        return count_coverage_indexed(samfile, chromosome_size=chromosome_size, flip=flip)\n","    all_counts = {}\n","    plus_strands = []\n","    minus_strands = []\n","    if \"SQ\" in samfile.header:\n","        chromosome_sizes = {}\n","        for entry in samfile.header[\"SQ\"]:\n","            chromosome_sizes[entry[\"SN\"]] = int(entry[\"LN\"]) + 1\n","    else:\n","        for reference in samfile.references:\n","            chromosome_sizes[reference] = chromosome_size\n","    for reference in samfile.references:  # create an array for each reference\n","        plus_strands.append(zeros((chromosome_sizes[reference],)))\n","        minus_strands.append(zeros((chromosome_sizes[reference],)))\n","    # iterate through each mapped read\n","    for i, read in enumerate(samfile):\n","        if read.is_unmapped:\n","            continue\n","        if not read.is_proper_pair:\n","            if read.is_reverse:\n","                minus_strands[read.tid][read.pos:read.aend] += 1\n","            else:\n","                plus_strands[read.tid][read.pos:read.aend] += 1\n","        # for paired-end data, get entire insert from only read1\n","        elif read.is_read1:\n","            if read.is_reverse:\n","                minus_strands[read.tid][read.pnext:read.aend] += 1\n","            else:\n","                plus_strands[read.tid][read.pos:read.pos + read.isize] += 1\n","    # store the results per reference\n","    for i, reference in enumerate(samfile.references):\n","        all_counts[reference] = {}\n","        if flip:\n","            all_counts[reference][\"-\"] = plus_strands[i]\n","            all_counts[reference][\"+\"] = minus_strands[i]\n","        else:\n","            all_counts[reference][\"+\"] = plus_strands[i]\n","            all_counts[reference][\"-\"] = minus_strands[i]\n","    return all_counts\n","\n","def count_coverage_indexed(samfile, chromosome_size=6000000, flip=True):\n","    \"\"\"counts coverage per base in a strand-specific manner\n","\n","    For paired-end reads, the insert betwen the mapped reads is\n","    also counted.\n","\n","    flip: Whether or not the strands should be flipped.\n","    This should be true for RNA-seq, and false for ChIP-exo\n","\n","    chromsome_size: This value should be larger than the largest chromosome\"\"\"\n","    if \"SQ\" in samfile.header:\n","        chromosome_sizes = {}\n","        for entry in samfile.header[\"SQ\"]:\n","            chromosome_sizes[entry[\"SN\"]] = int(entry[\"LN\"]) + 1\n","    else:\n","        for reference in samfile.references:\n","            chromosome_sizes[reference] = chromosome_size\n","    all_counts = {}\n","    for reference in samfile.references:  # go through each chromosome\n","        plus_strand = zeros((chromosome_sizes[reference],))\n","        minus_strand = zeros((chromosome_sizes[reference],))\n","        # iterate through each mapped read\n","        for i, read in enumerate(samfile.fetch(reference=reference)):\n","            if not read.is_proper_pair:\n","                if read.is_reverse:\n","                    minus_strand[read.pos:read.aend] += 1\n","                else:\n","                    plus_strand[read.pos:read.aend] += 1\n","            # for paired-end data, get entire insert from only read1\n","            elif read.is_read1:\n","                if read.is_reverse:\n","                    minus_strand[read.pnext:read.aend] += 1\n","                else:\n","                    plus_strand[read.pos:read.pos + read.isize] += 1\n","            all_counts[reference] = {}\n","            if flip:\n","                all_counts[reference][\"-\"] = plus_strand\n","                all_counts[reference][\"+\"] = minus_strand\n","            else:\n","                all_counts[reference][\"+\"] = plus_strand\n","                all_counts[reference][\"-\"] = minus_strand\n","    return all_counts\n","\n","def write_samfile_to_gff(samfile, output, chromosome_size=6000000,\n","                         separate_strand=True, flip=True, log_scale=True):\n","    \"\"\"write samfile object to an output object in a gff format\n","\n","    flip: Whether or not the strands should be flipped.\n","    This should be true for RNA-seq, and false for ChIP-exo\n","\n","    chromsome_size: This value should be larger than the largest chromosome\n","\n","    separate_strand: Whether the forward and reverse strands should be made\n","    into separate tracks (True) or the negative strand should be rendered\n","    as negative values (False)\n","    \"\"\"\n","    all_counts = count_coverage(samfile, chromosome_size=chromosome_size,\n","                                flip=flip)\n","\n","    name = os.path.split(samfile.filename)[1].decode()\n","    for reference in all_counts:\n","        for strand in all_counts[reference]:\n","            counts = all_counts[reference][strand]\n","            for i in counts.nonzero()[0]:\n","                if log_scale:\n","                    count = math.log(float(counts[i]), 2)\n","                else:\n","                    count = counts[i]\n","                if separate_strand:\n","                    output.write(\"%s\\t\\t%s\\t%d\\t%d\\t%.2f\\t%s\\t.\\t.\\n\" %\n","                        (reference, \"%s_(%s)\" % (name, strand), i, i,count, strand))\n","                else:\n","                    output.write(\"%s\\t\\t%s\\t%d\\t%d\\t%.2f\\t%s\\t.\\t.\\n\" %\n","                        (reference, name, i, i, count,strand))\n","\n","def convert_samfile_to_gff(sam_filename, out_filename, chromosome_size=6000000,\n","                           separate_strand=True, flip=True, log_scale=True):\n","    \"\"\"read in the a samfile from a path, and write it out to a gff filepath\n","\n","    flip: Whether or not the strands should be flipped.\n","    This should be true for RNA-seq, and false for ChIP-exo.\n","\n","    chromsome_size: This value should be larger than the largest chromosome.\n","\n","    separate_strand: Whether the forward and reverse strands should be made\n","    into separate tracks (True) or the negative strand should be rendered\n","    as negative values (False)\n","    \"\"\"\n","    samfile = pysam.Samfile(sam_filename)\n","    with open(out_filename, \"w\") as outfile:\n","        write_samfile_to_gff(samfile, outfile, chromosome_size=chromosome_size,\n","                             separate_strand=separate_strand, flip=flip,\n","                             log_scale=log_scale)\n","    samfile.close()\n","\n","def convert_sam_to_bedgraph(sam_bam_sorted, dir_bed_output):\n","    bedgraph_file_positive = f\"{dir_bed_output}_positive.bedgraph\"\n","    bedgraph_file_negative = f\"{dir_bed_output}_negative.bedgraph\"\n","\n","    !bedtools genomecov -ibam $sam_bam_sorted -bg -strand + > $bedgraph_file_positive\n","    !bedtools genomecov -ibam $sam_bam_sorted -bg -strand - > $bedgraph_file_negative"]},{"cell_type":"markdown","id":"6dae597d","metadata":{"id":"6dae597d"},"source":["### (Optional) Google Drive Mount\n","\n","If using Google Colab, it is necessary to mount Google Drive to access the files stored within it."]},{"cell_type":"code","execution_count":null,"id":"dd002bf3","metadata":{"id":"dd002bf3","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728289168360,"user_tz":-540,"elapsed":3180,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"e113d560-c123-4547-87a7-0fdb1f7232bc"},"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/mnt; to attempt to forcibly remount, call drive.mount(\"/content/mnt\", force_remount=True).\n"]}],"source":["from google.colab import drive, files\n","\n","drive.mount('/content/mnt')"]},{"cell_type":"markdown","id":"1bdb2c73","metadata":{"id":"1bdb2c73"},"source":["### Set the working directory"]},{"cell_type":"code","execution_count":null,"id":"04f2b429","metadata":{"id":"04f2b429","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728289170711,"user_tz":-540,"elapsed":425,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"d523031c-e725-405e-f0ea-3a386c3689a3"},"outputs":[{"output_type":"stream","name":"stdout","text":["Current working directiory : /content/mnt/My Drive/Test/RNA-seq\n"]}],"source":["working_directory =  '/content/mnt/My Drive/Test/RNA-seq'  #Enter your working directory.\n","\n","os.chdir(working_directory)\n","print(f'Current working directiory : {os.getcwd()}')"]},{"cell_type":"markdown","id":"773d7805","metadata":{"id":"773d7805"},"source":["## Step 2. Prepare the files\n","\n","Prepare the following files and move them into the working directory:\n","* **GenBank_file** : Enter the GenBank file name of the reference strain.\n","* **Organism** : Enter the accession number of the reference genome from the GenBank file. (e.g. *Escherichia coli* K-12 MG1655 : NC_000913.3)\n","* **Sample_name_list** : Enter the names of the prepared RNA-seq files, which include both R1 and R2 for paired-end sequencing, excluding the read number and file format (e.g., `RNA_seq_sample_1_R1.fastq.gz` and `RNA_seq_sample_1_R2.fastq.gz` -> `RNA_seq_sample_1`)."]},{"cell_type":"code","source":["GenBank_file = 'NC_000913.3.gb'\n","Organism = 'NC_000913.3'\n","Sample_name_list = ['mini_RNA-seq_Ecoli_mid-37-1',\n","                    'mini_RNA-seq_Ecoli_mid-37-2',\n","                    'mini_RNA-seq_Ecoli_del_rpoS_mid-37-1',\n","                    'mini_RNA-seq_Ecoli_del_rpoS_mid-37-2']"],"metadata":{"id":"q1Q5zVtloGIf"},"id":"q1Q5zVtloGIf","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"95ed683a","metadata":{"id":"95ed683a"},"source":["### Directory setting"]},{"cell_type":"code","execution_count":null,"id":"a4172915","metadata":{"id":"a4172915","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728289176190,"user_tz":-540,"elapsed":402,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"bbc58d05-538f-4635-a12d-fa74d01fc1bf"},"outputs":[{"output_type":"stream","name":"stdout","text":["Make directory : Success\n"]}],"source":["!mkdir '0_raw' '1_reference' '1_reference/rRNA_index' '1_reference/reference_genome_index' '2_alignment' '2_alignment/rRNA_del_fastq' '3_result'\n","!mv ./*.gb ./1_reference/\n","!mv ./*.gz ./0_raw/\n","print(f'Make directory : Success')"]},{"cell_type":"markdown","id":"05c859d6","metadata":{"id":"05c859d6"},"source":["## Step 3. Generate Fasta files from the reference genome file"]},{"cell_type":"code","execution_count":null,"id":"f07ba9f3","metadata":{"id":"f07ba9f3","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728289184978,"user_tz":-540,"elapsed":3656,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"5990007c-56f6-4dec-b78a-3eb191de6a14"},"outputs":[{"output_type":"stream","name":"stdout","text":["Step 3 is done.\n"]}],"source":["# Set the name of reference files.\n","dir_genbank_file = f'./1_reference/{GenBank_file}'\n","dir_genome_fasta = f'./1_reference/{Organism}.fasta'\n","dir_rRNA_fasta = f'./1_reference/{Organism}_rRNA.fasta'\n","\n","# Make the reference genome fasta file.\n","with open(dir_genome_fasta, \"w\") as genome_fasta:\n","    for record in SeqIO.parse(dir_genbank_file, \"genbank\"):\n","        SeqIO.write(record, genome_fasta, \"fasta\")\n","\n","# Make the ribosomal RNA of reference genome fasta file.\n","with open(dir_rRNA_fasta, \"w\") as rRNA_fasta:\n","    for record in SeqIO.parse(dir_genbank_file, \"genbank\"):\n","        for feature in record.features:\n","            if feature.type == \"rRNA\":\n","                rRNA_seq = feature.extract(record)\n","                rRNA_seq.id = Organism\n","                rRNA_seq.description = \"\"\n","                SeqIO.write(rRNA_seq, rRNA_fasta, \"fasta\")\n","\n","print('Step 3 is done.')"]},{"cell_type":"markdown","id":"04b9934d","metadata":{"id":"04b9934d"},"source":["## Step 4. Remove the rRNA sequences for RNA-seq dataset\n","\n","Unzip and align the sequencing reads file (.fastq.gz) to the rRNA sequences from reference genome using Bowtie 2 and remove the aligned rRNA reads from each sequencing read."]},{"cell_type":"code","execution_count":null,"id":"3c420cef","metadata":{"id":"3c420cef","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728289457618,"user_tz":-540,"elapsed":258745,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"d2942f70-2d2f-40c7-86d9-7488f272a122"},"outputs":[{"output_type":"stream","name":"stdout","text":["Settings:\n","  Output files: \"./1_reference/rRNA_index/NC_000913.3.*.bt2\"\n","  Line rate: 6 (line is 64 bytes)\n","  Lines per side: 1 (side is 64 bytes)\n","  Offset rate: 4 (one in 16)\n","  FTable chars: 10\n","  Strings: unpacked\n","  Max bucket size: default\n","  Max bucket size, sqrt multiplier: default\n","  Max bucket size, len divisor: 4\n","  Difference-cover sample period: 1024\n","  Endianness: little\n","  Actual local endianness: little\n","  Sanity checking: disabled\n","  Assertions: disabled\n","  Random seed: 0\n","  Sizeofs: void*:8, int:4, long:8, size_t:8\n","Input files DNA, FASTA:\n","  ./1_reference/NC_000913.3_rRNA.fasta\n","Building a SMALL index\n","Reading reference sizes\n","  Time reading reference sizes: 00:00:00\n","Calculating joined length\n","Writing header\n","Reserving space for joined string\n","Joining reference sequences\n","  Time to join reference sequences: 00:00:00\n","bmax according to bmaxDivN setting: 8020\n","Using parameters --bmax 6015 --dcv 1024\n","  Doing ahead-of-time memory usage test\n","  Passed!  Constructing with these parameters: --bmax 6015 --dcv 1024\n","Constructing suffix-array element generator\n","Building DifferenceCoverSample\n","  Building sPrime\n","  Building sPrimeOrder\n","  V-Sorting samples\n","  V-Sorting samples time: 00:00:00\n","  Allocating rank array\n","  Ranking v-sort output\n","  Ranking v-sort output time: 00:00:00\n","  Invoking Larsson-Sadakane on ranks\n","  Invoking Larsson-Sadakane on ranks time: 00:00:00\n","  Sanity-checking and returning\n","Building samples\n","Reserving space for 12 sample suffixes\n","Generating random suffixes\n","QSorting 12 sample offsets, eliminating duplicates\n","QSorting sample offsets, eliminating duplicates time: 00:00:00\n","Multikey QSorting 12 samples\n","  (Using difference cover)\n","  Multikey QSorting samples time: 00:00:00\n","Calculating bucket sizes\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Avg bucket size: 4009.5 (target: 6014)\n","Converting suffix-array elements to index image\n","Allocating ftab, absorbFtab\n","Entering Ebwt loop\n","Getting block 1 of 8\n","  Reserving size (6015) for bucket 1\n","  Calculating Z arrays for bucket 1\n","  Entering block accumulator loop for bucket 1:\n","  bucket 1: 10%\n","  bucket 1: 20%\n","  bucket 1: 30%\n","  bucket 1: 40%\n","  bucket 1: 50%\n","  bucket 1: 60%\n","  bucket 1: 70%\n","  bucket 1: 80%\n","  bucket 1: 90%\n","  bucket 1: 100%\n","  Sorting block of length 5287 for bucket 1\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5288 for bucket 1\n","Getting block 2 of 8\n","  Reserving size (6015) for bucket 2\n","  Calculating Z arrays for bucket 2\n","  Entering block accumulator loop for bucket 2:\n","  bucket 2: 10%\n","  bucket 2: 20%\n","  bucket 2: 30%\n","  bucket 2: 40%\n","  bucket 2: 50%\n","  bucket 2: 60%\n","  bucket 2: 70%\n","  bucket 2: 80%\n","  bucket 2: 90%\n","  bucket 2: 100%\n","  Sorting block of length 5768 for bucket 2\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5769 for bucket 2\n","Getting block 3 of 8\n","  Reserving size (6015) for bucket 3\n","  Calculating Z arrays for bucket 3\n","  Entering block accumulator loop for bucket 3:\n","  bucket 3: 10%\n","  bucket 3: 20%\n","  bucket 3: 30%\n","  bucket 3: 40%\n","  bucket 3: 50%\n","  bucket 3: 60%\n","  bucket 3: 70%\n","  bucket 3: 80%\n","  bucket 3: 90%\n","  bucket 3: 100%\n","  Sorting block of length 3072 for bucket 3\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 3073 for bucket 3\n","Getting block 4 of 8\n","  Reserving size (6015) for bucket 4\n","  Calculating Z arrays for bucket 4\n","  Entering block accumulator loop for bucket 4:\n","  bucket 4: 10%\n","  bucket 4: 20%\n","  bucket 4: 30%\n","  bucket 4: 40%\n","  bucket 4: 50%\n","  bucket 4: 60%\n","  bucket 4: 70%\n","  bucket 4: 80%\n","  bucket 4: 90%\n","  bucket 4: 100%\n","  Sorting block of length 4251 for bucket 4\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 4252 for bucket 4\n","Getting block 5 of 8\n","  Reserving size (6015) for bucket 5\n","  Calculating Z arrays for bucket 5\n","  Entering block accumulator loop for bucket 5:\n","  bucket 5: 10%\n","  bucket 5: 20%\n","  bucket 5: 30%\n","  bucket 5: 40%\n","  bucket 5: 50%\n","  bucket 5: 60%\n","  bucket 5: 70%\n","  bucket 5: 80%\n","  bucket 5: 90%\n","  bucket 5: 100%\n","  Sorting block of length 1943 for bucket 5\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 1944 for bucket 5\n","Getting block 6 of 8\n","  Reserving size (6015) for bucket 6\n","  Calculating Z arrays for bucket 6\n","  Entering block accumulator loop for bucket 6:\n","  bucket 6: 10%\n","  bucket 6: 20%\n","  bucket 6: 30%\n","  bucket 6: 40%\n","  bucket 6: 50%\n","  bucket 6: 60%\n","  bucket 6: 70%\n","  bucket 6: 80%\n","  bucket 6: 90%\n","  bucket 6: 100%\n","  Sorting block of length 5023 for bucket 6\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5024 for bucket 6\n","Getting block 7 of 8\n","  Reserving size (6015) for bucket 7\n","  Calculating Z arrays for bucket 7\n","  Entering block accumulator loop for bucket 7:\n","  bucket 7: 10%\n","  bucket 7: 20%\n","  bucket 7: 30%\n","  bucket 7: 40%\n","  bucket 7: 50%\n","  bucket 7: 60%\n","  bucket 7: 70%\n","  bucket 7: 80%\n","  bucket 7: 90%\n","  bucket 7: 100%\n","  Sorting block of length 2026 for bucket 7\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 2027 for bucket 7\n","Getting block 8 of 8\n","  Reserving size (6015) for bucket 8\n","  Calculating Z arrays for bucket 8\n","  Entering block accumulator loop for bucket 8:\n","  bucket 8: 10%\n","  bucket 8: 20%\n","  bucket 8: 30%\n","  bucket 8: 40%\n","  bucket 8: 50%\n","  bucket 8: 60%\n","  bucket 8: 70%\n","  bucket 8: 80%\n","  bucket 8: 90%\n","  bucket 8: 100%\n","  Sorting block of length 4706 for bucket 8\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 4707 for bucket 8\n","Exited Ebwt loop\n","fchr[A]: 0\n","fchr[C]: 8251\n","fchr[G]: 15465\n","fchr[T]: 25579\n","fchr[$]: 32083\n","Exiting Ebwt::buildToDisk()\n","Returning from initFromVector\n","Wrote 4205817 bytes to primary EBWT file: ./1_reference/rRNA_index/NC_000913.3.1.bt2\n","Wrote 8028 bytes to secondary EBWT file: ./1_reference/rRNA_index/NC_000913.3.2.bt2\n","Re-opening _in1 and _in2 as input streams\n","Returning from Ebwt constructor\n","Headers:\n","    len: 32083\n","    bwtLen: 32084\n","    sz: 8021\n","    bwtSz: 8021\n","    lineRate: 6\n","    offRate: 4\n","    offMask: 0xfffffff0\n","    ftabChars: 10\n","    eftabLen: 20\n","    eftabSz: 80\n","    ftabLen: 1048577\n","    ftabSz: 4194308\n","    offsLen: 2006\n","    offsSz: 8024\n","    lineSz: 64\n","    sideSz: 64\n","    sideBwtSz: 48\n","    sideBwtLen: 192\n","    numSides: 168\n","    numLines: 168\n","    ebwtTotLen: 10752\n","    ebwtTotSz: 10752\n","    color: 0\n","    reverse: 0\n","Total time for call to driver() for forward index: 00:00:01\n","Reading reference sizes\n","  Time reading reference sizes: 00:00:00\n","Calculating joined length\n","Writing header\n","Reserving space for joined string\n","Joining reference sequences\n","  Time to join reference sequences: 00:00:00\n","  Time to reverse reference sequence: 00:00:00\n","bmax according to bmaxDivN setting: 8020\n","Using parameters --bmax 6015 --dcv 1024\n","  Doing ahead-of-time memory usage test\n","  Passed!  Constructing with these parameters: --bmax 6015 --dcv 1024\n","Constructing suffix-array element generator\n","Building DifferenceCoverSample\n","  Building sPrime\n","  Building sPrimeOrder\n","  V-Sorting samples\n","  V-Sorting samples time: 00:00:00\n","  Allocating rank array\n","  Ranking v-sort output\n","  Ranking v-sort output time: 00:00:00\n","  Invoking Larsson-Sadakane on ranks\n","  Invoking Larsson-Sadakane on ranks time: 00:00:00\n","  Sanity-checking and returning\n","Building samples\n","Reserving space for 12 sample suffixes\n","Generating random suffixes\n","QSorting 12 sample offsets, eliminating duplicates\n","QSorting sample offsets, eliminating duplicates time: 00:00:00\n","Multikey QSorting 12 samples\n","  (Using difference cover)\n","  Multikey QSorting samples time: 00:00:00\n","Calculating bucket sizes\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Split 1, merged 7; iterating...\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Split 1, merged 1; iterating...\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Avg bucket size: 5346.33 (target: 6014)\n","Converting suffix-array elements to index image\n","Allocating ftab, absorbFtab\n","Entering Ebwt loop\n","Getting block 1 of 6\n","  Reserving size (6015) for bucket 1\n","  Calculating Z arrays for bucket 1\n","  Entering block accumulator loop for bucket 1:\n","  bucket 1: 10%\n","  bucket 1: 20%\n","  bucket 1: 30%\n","  bucket 1: 40%\n","  bucket 1: 50%\n","  bucket 1: 60%\n","  bucket 1: 70%\n","  bucket 1: 80%\n","  bucket 1: 90%\n","  bucket 1: 100%\n","  Sorting block of length 5197 for bucket 1\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5198 for bucket 1\n","Getting block 2 of 6\n","  Reserving size (6015) for bucket 2\n","  Calculating Z arrays for bucket 2\n","  Entering block accumulator loop for bucket 2:\n","  bucket 2: 10%\n","  bucket 2: 20%\n","  bucket 2: 30%\n","  bucket 2: 40%\n","  bucket 2: 50%\n","  bucket 2: 60%\n","  bucket 2: 70%\n","  bucket 2: 80%\n","  bucket 2: 90%\n","  bucket 2: 100%\n","  Sorting block of length 5158 for bucket 2\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5159 for bucket 2\n","Getting block 3 of 6\n","  Reserving size (6015) for bucket 3\n","  Calculating Z arrays for bucket 3\n","  Entering block accumulator loop for bucket 3:\n","  bucket 3: 10%\n","  bucket 3: 20%\n","  bucket 3: 30%\n","  bucket 3: 40%\n","  bucket 3: 50%\n","  bucket 3: 60%\n","  bucket 3: 70%\n","  bucket 3: 80%\n","  bucket 3: 90%\n","  bucket 3: 100%\n","  Sorting block of length 4893 for bucket 3\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 4894 for bucket 3\n","Getting block 4 of 6\n","  Reserving size (6015) for bucket 4\n","  Calculating Z arrays for bucket 4\n","  Entering block accumulator loop for bucket 4:\n","  bucket 4: 10%\n","  bucket 4: 20%\n","  bucket 4: 30%\n","  bucket 4: 40%\n","  bucket 4: 50%\n","  bucket 4: 60%\n","  bucket 4: 70%\n","  bucket 4: 80%\n","  bucket 4: 90%\n","  bucket 4: 100%\n","  Sorting block of length 5751 for bucket 4\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5752 for bucket 4\n","Getting block 5 of 6\n","  Reserving size (6015) for bucket 5\n","  Calculating Z arrays for bucket 5\n","  Entering block accumulator loop for bucket 5:\n","  bucket 5: 10%\n","  bucket 5: 20%\n","  bucket 5: 30%\n","  bucket 5: 40%\n","  bucket 5: 50%\n","  bucket 5: 60%\n","  bucket 5: 70%\n","  bucket 5: 80%\n","  bucket 5: 90%\n","  bucket 5: 100%\n","  Sorting block of length 5975 for bucket 5\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5976 for bucket 5\n","Getting block 6 of 6\n","  Reserving size (6015) for bucket 6\n","  Calculating Z arrays for bucket 6\n","  Entering block accumulator loop for bucket 6:\n","  bucket 6: 10%\n","  bucket 6: 20%\n","  bucket 6: 30%\n","  bucket 6: 40%\n","  bucket 6: 50%\n","  bucket 6: 60%\n","  bucket 6: 70%\n","  bucket 6: 80%\n","  bucket 6: 90%\n","  bucket 6: 100%\n","  Sorting block of length 5104 for bucket 6\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 5105 for bucket 6\n","Exited Ebwt loop\n","fchr[A]: 0\n","fchr[C]: 8251\n","fchr[G]: 15465\n","fchr[T]: 25579\n","fchr[$]: 32083\n","Exiting Ebwt::buildToDisk()\n","Returning from initFromVector\n","Wrote 4205817 bytes to primary EBWT file: ./1_reference/rRNA_index/NC_000913.3.rev.1.bt2\n","Wrote 8028 bytes to secondary EBWT file: ./1_reference/rRNA_index/NC_000913.3.rev.2.bt2\n","Re-opening _in1 and _in2 as input streams\n","Returning from Ebwt constructor\n","Headers:\n","    len: 32083\n","    bwtLen: 32084\n","    sz: 8021\n","    bwtSz: 8021\n","    lineRate: 6\n","    offRate: 4\n","    offMask: 0xfffffff0\n","    ftabChars: 10\n","    eftabLen: 20\n","    eftabSz: 80\n","    ftabLen: 1048577\n","    ftabSz: 4194308\n","    offsLen: 2006\n","    offsSz: 8024\n","    lineSz: 64\n","    sideSz: 64\n","    sideBwtSz: 48\n","    sideBwtLen: 192\n","    numSides: 168\n","    numLines: 168\n","    ebwtTotLen: 10752\n","    ebwtTotSz: 10752\n","    color: 0\n","    reverse: 1\n","Total time for backward call to driver() for mirror index: 00:00:00\n","mini_RNA-seq_Ecoli_mid-37-1 processing ...\n","\t Aligning to rRNA sequences of reference genome...\n","239373 reads; of these:\n","  239373 (100.00%) were paired; of these:\n","    227950 (95.23%) aligned concordantly 0 times\n","    10 (0.00%) aligned concordantly exactly 1 time\n","    11413 (4.77%) aligned concordantly >1 times\n","    ----\n","    227950 pairs aligned concordantly 0 times; of these:\n","      0 (0.00%) aligned discordantly 1 time\n","    ----\n","    227950 pairs aligned 0 times concordantly or discordantly; of these:\n","      455900 mates make up the pairs; of these:\n","        455476 (99.91%) aligned 0 times\n","        4 (0.00%) aligned exactly 1 time\n","        420 (0.09%) aligned >1 times\n","4.86% overall alignment rate\n","mini_RNA-seq_Ecoli_mid-37-2 processing ...\n","\t Aligning to rRNA sequences of reference genome...\n","244833 reads; of these:\n","  244833 (100.00%) were paired; of these:\n","    232210 (94.84%) aligned concordantly 0 times\n","    7 (0.00%) aligned concordantly exactly 1 time\n","    12616 (5.15%) aligned concordantly >1 times\n","    ----\n","    232210 pairs aligned concordantly 0 times; of these:\n","      0 (0.00%) aligned discordantly 1 time\n","    ----\n","    232210 pairs aligned 0 times concordantly or discordantly; of these:\n","      464420 mates make up the pairs; of these:\n","        463872 (99.88%) aligned 0 times\n","        5 (0.00%) aligned exactly 1 time\n","        543 (0.12%) aligned >1 times\n","5.27% overall alignment rate\n","mini_RNA-seq_Ecoli_del_rpoS_mid-37-1 processing ...\n","\t Aligning to rRNA sequences of reference genome...\n","284663 reads; of these:\n","  284663 (100.00%) were paired; of these:\n","    263701 (92.64%) aligned concordantly 0 times\n","    10 (0.00%) aligned concordantly exactly 1 time\n","    20952 (7.36%) aligned concordantly >1 times\n","    ----\n","    263701 pairs aligned concordantly 0 times; of these:\n","      0 (0.00%) aligned discordantly 1 time\n","    ----\n","    263701 pairs aligned 0 times concordantly or discordantly; of these:\n","      527402 mates make up the pairs; of these:\n","        526737 (99.87%) aligned 0 times\n","        6 (0.00%) aligned exactly 1 time\n","        659 (0.12%) aligned >1 times\n","7.48% overall alignment rate\n","mini_RNA-seq_Ecoli_del_rpoS_mid-37-2 processing ...\n","\t Aligning to rRNA sequences of reference genome...\n","346838 reads; of these:\n","  346838 (100.00%) were paired; of these:\n","    327715 (94.49%) aligned concordantly 0 times\n","    7 (0.00%) aligned concordantly exactly 1 time\n","    19116 (5.51%) aligned concordantly >1 times\n","    ----\n","    327715 pairs aligned concordantly 0 times; of these:\n","      0 (0.00%) aligned discordantly 1 time\n","    ----\n","    327715 pairs aligned 0 times concordantly or discordantly; of these:\n","      655430 mates make up the pairs; of these:\n","        654771 (99.90%) aligned 0 times\n","        8 (0.00%) aligned exactly 1 time\n","        651 (0.10%) aligned >1 times\n","5.61% overall alignment rate\n","Step 4 is done.\n"]}],"source":["# Build the bowtie2 index set for removing the rRNA sequences\n","dir_rRNA_ref = f'./1_reference/rRNA_index/{Organism}'\n","!bowtie2-build $dir_rRNA_fasta $dir_rRNA_ref\n","\n","#Data pre-process\n","for file in Sample_name_list:\n","    print(f'{file} processing ...')\n","\n","    raw_R1_gz = f'./0_raw/{file}_R1.fastq.gz'\n","    raw_R2_gz = f'./0_raw/{file}_R2.fastq.gz'\n","\n","    # Unzip the fastq.gz file\n","    !gunzip -k $raw_R1_gz > /dev/null\n","    !gunzip -k $raw_R2_gz > /dev/null\n","\n","    raw_R1_fastq = f'./0_raw/{file}_R1.fastq'\n","    raw_R2_fastq = f'./0_raw/{file}_R2.fastq'\n","\n","    rRNA_del_fastq_gz = f'./2_alignment/rRNA_del_fastq/rRNA_del_{file}.fastq.gz'\n","    rRNA_mapped_sam = f'./2_alignment/rRNA_mapped_{file}.sam'\n","\n","    print('\\t Aligning to rRNA sequences of reference genome...')\n","    !bowtie2 -x $dir_rRNA_ref -1 $raw_R1_fastq -2 $raw_R2_fastq --un-conc-gz $rRNA_del_fastq_gz -S $rRNA_mapped_sam\n","    !rm $rRNA_mapped_sam\n","\n","print('Step 4 is done.')"]},{"cell_type":"markdown","id":"c131b070","metadata":{"id":"c131b070"},"source":["## Step 5. Align the RNA-seq reads to reference genome sequence\n","\n","Align the rRNA-removed sequencing reads from step 3 to reference genome using Bowtie 2, with a maximum insert size of 1000 bp, allowing 1 mismatch after trimming 3 bp at 3' ends."]},{"cell_type":"code","execution_count":null,"id":"d8224c8c","metadata":{"id":"d8224c8c","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728289732412,"user_tz":-540,"elapsed":274801,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"8c9d166c-5284-4f9e-cdd6-00628e4332b7"},"outputs":[{"output_type":"stream","name":"stdout","text":["Settings:\n","  Output files: \"./1_reference/reference_genome_index/NC_000913.3.*.bt2\"\n","  Line rate: 6 (line is 64 bytes)\n","  Lines per side: 1 (side is 64 bytes)\n","  Offset rate: 4 (one in 16)\n","  FTable chars: 10\n","  Strings: unpacked\n","  Max bucket size: default\n","  Max bucket size, sqrt multiplier: default\n","  Max bucket size, len divisor: 4\n","  Difference-cover sample period: 1024\n","  Endianness: little\n","  Actual local endianness: little\n","  Sanity checking: disabled\n","  Assertions: disabled\n","  Random seed: 0\n","  Sizeofs: void*:8, int:4, long:8, size_t:8\n","Input files DNA, FASTA:\n","  ./1_reference/NC_000913.3.fasta\n","Building a SMALL index\n","Reading reference sizes\n","  Time reading reference sizes: 00:00:00\n","Calculating joined length\n","Writing header\n","Reserving space for joined string\n","Joining reference sequences\n","  Time to join reference sequences: 00:00:00\n","bmax according to bmaxDivN setting: 1160413\n","Using parameters --bmax 870310 --dcv 1024\n","  Doing ahead-of-time memory usage test\n","  Passed!  Constructing with these parameters: --bmax 870310 --dcv 1024\n","Constructing suffix-array element generator\n","Building DifferenceCoverSample\n","  Building sPrime\n","  Building sPrimeOrder\n","  V-Sorting samples\n","  V-Sorting samples time: 00:00:00\n","  Allocating rank array\n","  Ranking v-sort output\n","  Ranking v-sort output time: 00:00:00\n","  Invoking Larsson-Sadakane on ranks\n","  Invoking Larsson-Sadakane on ranks time: 00:00:00\n","  Sanity-checking and returning\n","Building samples\n","Reserving space for 12 sample suffixes\n","Generating random suffixes\n","QSorting 12 sample offsets, eliminating duplicates\n","QSorting sample offsets, eliminating duplicates time: 00:00:00\n","Multikey QSorting 12 samples\n","  (Using difference cover)\n","  Multikey QSorting samples time: 00:00:00\n","Calculating bucket sizes\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Split 1, merged 6; iterating...\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Avg bucket size: 663092 (target: 870309)\n","Converting suffix-array elements to index image\n","Allocating ftab, absorbFtab\n","Entering Ebwt loop\n","Getting block 1 of 7\n","  Reserving size (870310) for bucket 1\n","  Calculating Z arrays for bucket 1\n","  Entering block accumulator loop for bucket 1:\n","  bucket 1: 10%\n","  bucket 1: 20%\n","  bucket 1: 30%\n","  bucket 1: 40%\n","  bucket 1: 50%\n","  bucket 1: 60%\n","  bucket 1: 70%\n","  bucket 1: 80%\n","  bucket 1: 90%\n","  bucket 1: 100%\n","  Sorting block of length 415161 for bucket 1\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 415162 for bucket 1\n","Getting block 2 of 7\n","  Reserving size (870310) for bucket 2\n","  Calculating Z arrays for bucket 2\n","  Entering block accumulator loop for bucket 2:\n","  bucket 2: 10%\n","  bucket 2: 20%\n","  bucket 2: 30%\n","  bucket 2: 40%\n","  bucket 2: 50%\n","  bucket 2: 60%\n","  bucket 2: 70%\n","  bucket 2: 80%\n","  bucket 2: 90%\n","  bucket 2: 100%\n","  Sorting block of length 704345 for bucket 2\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 704346 for bucket 2\n","Getting block 3 of 7\n","  Reserving size (870310) for bucket 3\n","  Calculating Z arrays for bucket 3\n","  Entering block accumulator loop for bucket 3:\n","  bucket 3: 10%\n","  bucket 3: 20%\n","  bucket 3: 30%\n","  bucket 3: 40%\n","  bucket 3: 50%\n","  bucket 3: 60%\n","  bucket 3: 70%\n","  bucket 3: 80%\n","  bucket 3: 90%\n","  bucket 3: 100%\n","  Sorting block of length 715232 for bucket 3\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 715233 for bucket 3\n","Getting block 4 of 7\n","  Reserving size (870310) for bucket 4\n","  Calculating Z arrays for bucket 4\n","  Entering block accumulator loop for bucket 4:\n","  bucket 4: 10%\n","  bucket 4: 20%\n","  bucket 4: 30%\n","  bucket 4: 40%\n","  bucket 4: 50%\n","  bucket 4: 60%\n","  bucket 4: 70%\n","  bucket 4: 80%\n","  bucket 4: 90%\n","  bucket 4: 100%\n","  Sorting block of length 638686 for bucket 4\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 638687 for bucket 4\n","Getting block 5 of 7\n","  Reserving size (870310) for bucket 5\n","  Calculating Z arrays for bucket 5\n","  Entering block accumulator loop for bucket 5:\n","  bucket 5: 10%\n","  bucket 5: 20%\n","  bucket 5: 30%\n","  bucket 5: 40%\n","  bucket 5: 50%\n","  bucket 5: 60%\n","  bucket 5: 70%\n","  bucket 5: 80%\n","  bucket 5: 90%\n","  bucket 5: 100%\n","  Sorting block of length 779115 for bucket 5\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 779116 for bucket 5\n","Getting block 6 of 7\n","  Reserving size (870310) for bucket 6\n","  Calculating Z arrays for bucket 6\n","  Entering block accumulator loop for bucket 6:\n","  bucket 6: 10%\n","  bucket 6: 20%\n","  bucket 6: 30%\n","  bucket 6: 40%\n","  bucket 6: 50%\n","  bucket 6: 60%\n","  bucket 6: 70%\n","  bucket 6: 80%\n","  bucket 6: 90%\n","  bucket 6: 100%\n","  Sorting block of length 548871 for bucket 6\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 548872 for bucket 6\n","Getting block 7 of 7\n","  Reserving size (870310) for bucket 7\n","  Calculating Z arrays for bucket 7\n","  Entering block accumulator loop for bucket 7:\n","  bucket 7: 10%\n","  bucket 7: 20%\n","  bucket 7: 30%\n","  bucket 7: 40%\n","  bucket 7: 50%\n","  bucket 7: 60%\n","  bucket 7: 70%\n","  bucket 7: 80%\n","  bucket 7: 90%\n","  bucket 7: 100%\n","  Sorting block of length 840236 for bucket 7\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 840237 for bucket 7\n","Exited Ebwt loop\n","fchr[A]: 0\n","fchr[C]: 1142742\n","fchr[G]: 2322833\n","fchr[T]: 3500270\n","fchr[$]: 4641652\n","Exiting Ebwt::buildToDisk()\n","Returning from initFromVector\n","Wrote 5741800 bytes to primary EBWT file: ./1_reference/reference_genome_index/NC_000913.3.1.bt2\n","Wrote 1160420 bytes to secondary EBWT file: ./1_reference/reference_genome_index/NC_000913.3.2.bt2\n","Re-opening _in1 and _in2 as input streams\n","Returning from Ebwt constructor\n","Headers:\n","    len: 4641652\n","    bwtLen: 4641653\n","    sz: 1160413\n","    bwtSz: 1160414\n","    lineRate: 6\n","    offRate: 4\n","    offMask: 0xfffffff0\n","    ftabChars: 10\n","    eftabLen: 20\n","    eftabSz: 80\n","    ftabLen: 1048577\n","    ftabSz: 4194308\n","    offsLen: 290104\n","    offsSz: 1160416\n","    lineSz: 64\n","    sideSz: 64\n","    sideBwtSz: 48\n","    sideBwtLen: 192\n","    numSides: 24176\n","    numLines: 24176\n","    ebwtTotLen: 1547264\n","    ebwtTotSz: 1547264\n","    color: 0\n","    reverse: 0\n","Total time for call to driver() for forward index: 00:00:03\n","Reading reference sizes\n","  Time reading reference sizes: 00:00:00\n","Calculating joined length\n","Writing header\n","Reserving space for joined string\n","Joining reference sequences\n","  Time to join reference sequences: 00:00:00\n","  Time to reverse reference sequence: 00:00:00\n","bmax according to bmaxDivN setting: 1160413\n","Using parameters --bmax 870310 --dcv 1024\n","  Doing ahead-of-time memory usage test\n","  Passed!  Constructing with these parameters: --bmax 870310 --dcv 1024\n","Constructing suffix-array element generator\n","Building DifferenceCoverSample\n","  Building sPrime\n","  Building sPrimeOrder\n","  V-Sorting samples\n","  V-Sorting samples time: 00:00:00\n","  Allocating rank array\n","  Ranking v-sort output\n","  Ranking v-sort output time: 00:00:00\n","  Invoking Larsson-Sadakane on ranks\n","  Invoking Larsson-Sadakane on ranks time: 00:00:00\n","  Sanity-checking and returning\n","Building samples\n","Reserving space for 12 sample suffixes\n","Generating random suffixes\n","QSorting 12 sample offsets, eliminating duplicates\n","QSorting sample offsets, eliminating duplicates time: 00:00:00\n","Multikey QSorting 12 samples\n","  (Using difference cover)\n","  Multikey QSorting samples time: 00:00:00\n","Calculating bucket sizes\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Split 2, merged 7; iterating...\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Split 1, merged 1; iterating...\n","Splitting and merging\n","  Splitting and merging time: 00:00:00\n","Avg bucket size: 580206 (target: 870309)\n","Converting suffix-array elements to index image\n","Allocating ftab, absorbFtab\n","Entering Ebwt loop\n","Getting block 1 of 8\n","  Reserving size (870310) for bucket 1\n","  Calculating Z arrays for bucket 1\n","  Entering block accumulator loop for bucket 1:\n","  bucket 1: 10%\n","  bucket 1: 20%\n","  bucket 1: 30%\n","  bucket 1: 40%\n","  bucket 1: 50%\n","  bucket 1: 60%\n","  bucket 1: 70%\n","  bucket 1: 80%\n","  bucket 1: 90%\n","  bucket 1: 100%\n","  Sorting block of length 696384 for bucket 1\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 696385 for bucket 1\n","Getting block 2 of 8\n","  Reserving size (870310) for bucket 2\n","  Calculating Z arrays for bucket 2\n","  Entering block accumulator loop for bucket 2:\n","  bucket 2: 10%\n","  bucket 2: 20%\n","  bucket 2: 30%\n","  bucket 2: 40%\n","  bucket 2: 50%\n","  bucket 2: 60%\n","  bucket 2: 70%\n","  bucket 2: 80%\n","  bucket 2: 90%\n","  bucket 2: 100%\n","  Sorting block of length 333881 for bucket 2\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 333882 for bucket 2\n","Getting block 3 of 8\n","  Reserving size (870310) for bucket 3\n","  Calculating Z arrays for bucket 3\n","  Entering block accumulator loop for bucket 3:\n","  bucket 3: 10%\n","  bucket 3: 20%\n","  bucket 3: 30%\n","  bucket 3: 40%\n","  bucket 3: 50%\n","  bucket 3: 60%\n","  bucket 3: 70%\n","  bucket 3: 80%\n","  bucket 3: 90%\n","  bucket 3: 100%\n","  Sorting block of length 565798 for bucket 3\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 565799 for bucket 3\n","Getting block 4 of 8\n","  Reserving size (870310) for bucket 4\n","  Calculating Z arrays for bucket 4\n","  Entering block accumulator loop for bucket 4:\n","  bucket 4: 10%\n","  bucket 4: 20%\n","  bucket 4: 30%\n","  bucket 4: 40%\n","  bucket 4: 50%\n","  bucket 4: 60%\n","  bucket 4: 70%\n","  bucket 4: 80%\n","  bucket 4: 90%\n","  bucket 4: 100%\n","  Sorting block of length 312583 for bucket 4\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 312584 for bucket 4\n","Getting block 5 of 8\n","  Reserving size (870310) for bucket 5\n","  Calculating Z arrays for bucket 5\n","  Entering block accumulator loop for bucket 5:\n","  bucket 5: 10%\n","  bucket 5: 20%\n","  bucket 5: 30%\n","  bucket 5: 40%\n","  bucket 5: 50%\n","  bucket 5: 60%\n","  bucket 5: 70%\n","  bucket 5: 80%\n","  bucket 5: 90%\n","  bucket 5: 100%\n","  Sorting block of length 783371 for bucket 5\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 783372 for bucket 5\n","Getting block 6 of 8\n","  Reserving size (870310) for bucket 6\n","  Calculating Z arrays for bucket 6\n","  Entering block accumulator loop for bucket 6:\n","  bucket 6: 10%\n","  bucket 6: 20%\n","  bucket 6: 30%\n","  bucket 6: 40%\n","  bucket 6: 50%\n","  bucket 6: 60%\n","  bucket 6: 70%\n","  bucket 6: 80%\n","  bucket 6: 90%\n","  bucket 6: 100%\n","  Sorting block of length 823242 for bucket 6\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 823243 for bucket 6\n","Getting block 7 of 8\n","  Reserving size (870310) for bucket 7\n","  Calculating Z arrays for bucket 7\n","  Entering block accumulator loop for bucket 7:\n","  bucket 7: 10%\n","  bucket 7: 20%\n","  bucket 7: 30%\n","  bucket 7: 40%\n","  bucket 7: 50%\n","  bucket 7: 60%\n","  bucket 7: 70%\n","  bucket 7: 80%\n","  bucket 7: 90%\n","  bucket 7: 100%\n","  Sorting block of length 609090 for bucket 7\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 609091 for bucket 7\n","Getting block 8 of 8\n","  Reserving size (870310) for bucket 8\n","  Calculating Z arrays for bucket 8\n","  Entering block accumulator loop for bucket 8:\n","  bucket 8: 10%\n","  bucket 8: 20%\n","  bucket 8: 30%\n","  bucket 8: 40%\n","  bucket 8: 50%\n","  bucket 8: 60%\n","  bucket 8: 70%\n","  bucket 8: 80%\n","  bucket 8: 90%\n","  bucket 8: 100%\n","  Sorting block of length 517296 for bucket 8\n","  (Using difference cover)\n","  Sorting block time: 00:00:00\n","Returning block of 517297 for bucket 8\n","Exited Ebwt loop\n","fchr[A]: 0\n","fchr[C]: 1142742\n","fchr[G]: 2322833\n","fchr[T]: 3500270\n","fchr[$]: 4641652\n","Exiting Ebwt::buildToDisk()\n","Returning from initFromVector\n","Wrote 5741800 bytes to primary EBWT file: ./1_reference/reference_genome_index/NC_000913.3.rev.1.bt2\n","Wrote 1160420 bytes to secondary EBWT file: ./1_reference/reference_genome_index/NC_000913.3.rev.2.bt2\n","Re-opening _in1 and _in2 as input streams\n","Returning from Ebwt constructor\n","Headers:\n","    len: 4641652\n","    bwtLen: 4641653\n","    sz: 1160413\n","    bwtSz: 1160414\n","    lineRate: 6\n","    offRate: 4\n","    offMask: 0xfffffff0\n","    ftabChars: 10\n","    eftabLen: 20\n","    eftabSz: 80\n","    ftabLen: 1048577\n","    ftabSz: 4194308\n","    offsLen: 290104\n","    offsSz: 1160416\n","    lineSz: 64\n","    sideSz: 64\n","    sideBwtSz: 48\n","    sideBwtLen: 192\n","    numSides: 24176\n","    numLines: 24176\n","    ebwtTotLen: 1547264\n","    ebwtTotSz: 1547264\n","    color: 0\n","    reverse: 1\n","Total time for backward call to driver() for mirror index: 00:00:04\n","mini_RNA-seq_Ecoli_mid-37-1 processing ...\n","\t Aligning to reference genome...\n","227950 reads; of these:\n","  227950 (100.00%) were paired; of these:\n","    6132 (2.69%) aligned concordantly 0 times\n","    214503 (94.10%) aligned concordantly exactly 1 time\n","    7315 (3.21%) aligned concordantly >1 times\n","    ----\n","    6132 pairs aligned concordantly 0 times; of these:\n","      236 (3.85%) aligned discordantly 1 time\n","    ----\n","    5896 pairs aligned 0 times concordantly or discordantly; of these:\n","      11792 mates make up the pairs; of these:\n","        9704 (82.29%) aligned 0 times\n","        1819 (15.43%) aligned exactly 1 time\n","        269 (2.28%) aligned >1 times\n","97.87% overall alignment rate\n","mini_RNA-seq_Ecoli_mid-37-2 processing ...\n","\t Aligning to reference genome...\n","232210 reads; of these:\n","  232210 (100.00%) were paired; of these:\n","    3751 (1.62%) aligned concordantly 0 times\n","    220960 (95.16%) aligned concordantly exactly 1 time\n","    7499 (3.23%) aligned concordantly >1 times\n","    ----\n","    3751 pairs aligned concordantly 0 times; of these:\n","      320 (8.53%) aligned discordantly 1 time\n","    ----\n","    3431 pairs aligned 0 times concordantly or discordantly; of these:\n","      6862 mates make up the pairs; of these:\n","        4853 (70.72%) aligned 0 times\n","        1752 (25.53%) aligned exactly 1 time\n","        257 (3.75%) aligned >1 times\n","98.96% overall alignment rate\n","mini_RNA-seq_Ecoli_del_rpoS_mid-37-1 processing ...\n","\t Aligning to reference genome...\n","263701 reads; of these:\n","  263701 (100.00%) were paired; of these:\n","    5623 (2.13%) aligned concordantly 0 times\n","    249198 (94.50%) aligned concordantly exactly 1 time\n","    8880 (3.37%) aligned concordantly >1 times\n","    ----\n","    5623 pairs aligned concordantly 0 times; of these:\n","      359 (6.38%) aligned discordantly 1 time\n","    ----\n","    5264 pairs aligned 0 times concordantly or discordantly; of these:\n","      10528 mates make up the pairs; of these:\n","        7648 (72.64%) aligned 0 times\n","        2447 (23.24%) aligned exactly 1 time\n","        433 (4.11%) aligned >1 times\n","98.55% overall alignment rate\n","mini_RNA-seq_Ecoli_del_rpoS_mid-37-2 processing ...\n","\t Aligning to reference genome...\n","327715 reads; of these:\n","  327715 (100.00%) were paired; of these:\n","    5914 (1.80%) aligned concordantly 0 times\n","    310906 (94.87%) aligned concordantly exactly 1 time\n","    10895 (3.32%) aligned concordantly >1 times\n","    ----\n","    5914 pairs aligned concordantly 0 times; of these:\n","      495 (8.37%) aligned discordantly 1 time\n","    ----\n","    5419 pairs aligned 0 times concordantly or discordantly; of these:\n","      10838 mates make up the pairs; of these:\n","        7690 (70.95%) aligned 0 times\n","        2727 (25.16%) aligned exactly 1 time\n","        421 (3.88%) aligned >1 times\n","98.83% overall alignment rate\n","Step 5 is done.\n"]}],"source":["# Build the bowtie2 index set of reference genome\n","dir_genome_ref = f'./1_reference/reference_genome_index/{Organism}'\n","!bowtie2-build $dir_genome_fasta $dir_genome_ref\n","\n","for file in Sample_name_list:\n","    print(f'{file} processing ...')\n","\n","    rRNA_del_R1_fastq_gz = f'./2_alignment/rRNA_del_fastq/rRNA_del_{file}.fastq.1.gz'\n","    rRNA_del_R2_fastq_gz = f'./2_alignment/rRNA_del_fastq/rRNA_del_{file}.fastq.2.gz'\n","    re_rRNA_del_R1_fastq_gz = f'./2_alignment/rRNA_del_fastq/rRNA_del_{file}_R1.fastq.gz'\n","    re_rRNA_del_R2_fastq_gz = f'./2_alignment/rRNA_del_fastq/rRNA_del_{file}_R2.fastq.gz'\n","\n","    os.rename(rRNA_del_R1_fastq_gz,re_rRNA_del_R1_fastq_gz)\n","    os.rename(rRNA_del_R2_fastq_gz,re_rRNA_del_R2_fastq_gz)\n","\n","    !gunzip -k $re_rRNA_del_R1_fastq_gz > /dev/null\n","    !gunzip -k $re_rRNA_del_R2_fastq_gz > /dev/null\n","\n","    R1_file = re_rRNA_del_R1_fastq_gz.split(\".gz\")[0]\n","    R2_file = re_rRNA_del_R2_fastq_gz.split(\".gz\")[0]\n","\n","    genome_mapped_sam = f'./2_alignment/genome_mapped_{file}.sam'\n","\n","    print('\\t Aligning to reference genome...')\n","    !bowtie2 -x $dir_genome_ref -1 $R1_file -2 $R2_file -S $genome_mapped_sam -X 1000 -N 1 -3 3\n","\n","print('Step 5 is done.')"]},{"cell_type":"markdown","id":"e316178c","metadata":{"id":"e316178c"},"source":["## Step 6. Convert the SAM files generated from Bowtie 2 into BAM, GFF, and Bedgraph files\n","\n","Convert the SAM files generated from Bowtie 2 into BAM using Samtools and generate GFF files for downstream analysis."]},{"cell_type":"code","execution_count":null,"id":"bc72f853","metadata":{"id":"bc72f853","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728290032858,"user_tz":-540,"elapsed":300454,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"b24e5b44-3875-43b7-b405-ee1c6821c81e"},"outputs":[{"output_type":"stream","name":"stdout","text":["mini_RNA-seq_Ecoli_mid-37-1 processing ...\n","\t Samtools running is done.\n","\t Making the gff file is done.\n","\t Making the bedGrpah file is done.\n","\n","mini_RNA-seq_Ecoli_mid-37-2 processing ...\n","\t Samtools running is done.\n","\t Making the gff file is done.\n","\t Making the bedGrpah file is done.\n","\n","mini_RNA-seq_Ecoli_del_rpoS_mid-37-1 processing ...\n","\t Samtools running is done.\n","\t Making the gff file is done.\n","\t Making the bedGrpah file is done.\n","\n","mini_RNA-seq_Ecoli_del_rpoS_mid-37-2 processing ...\n","\t Samtools running is done.\n","\t Making the gff file is done.\n","\t Making the bedGrpah file is done.\n","\n","Step 6 is done.\n"]}],"source":["for file in Sample_name_list:\n","    print(f'{file} processing ...')\n","    sam_bam_unsorted = f'./2_alignment/{file}.unsorted.bam'\n","    sam_bam_sorted = f'./2_alignment/{file}.bam'\n","    genome_mapped_sam = f'./2_alignment/genome_mapped_{file}.sam'\n","\n","    !samtools view -bS $genome_mapped_sam -o $sam_bam_unsorted\n","    !samtools sort $sam_bam_unsorted -o $sam_bam_sorted\n","    !rm $sam_bam_unsorted\n","    !samtools index $sam_bam_sorted\n","\n","    print(\"\\t Samtools running is done.\")\n","\n","    dir_gff_output = f'./3_result/{file}.gff'\n","    convert_samfile_to_gff( genome_mapped_sam, dir_gff_output )\n","    print(\"\\t Making the gff file is done.\")\n","\n","    dir_bed_output = f'./3_result/{file}'\n","    convert_sam_to_bedgraph(genome_mapped_sam, dir_bed_output)\n","    print('\\t Making the bedGrpah file is done.\\n')\n","\n","print('Step 6 is done.')"]},{"cell_type":"code","source":[],"metadata":{"id":"D0i9QTQbuqps"},"id":"D0i9QTQbuqps","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}