{"cells":[{"cell_type":"markdown","metadata":{"id":"4T2GvMCtm8uI"},"source":["# R-based pipeline for calculating the DEGs using DESeq2"]},{"cell_type":"markdown","metadata":{"id":"_YXpnJ2fm8uR"},"source":["## Protocol B-2 - Calculation of Differentially Expressed Genes (DEGs) for RNA-seq  dataset"]},{"cell_type":"markdown","metadata":{"id":"7PQCbFCHm8uS"},"source":["## Step 1. Set-up\n","\n","### Install and import libraries\n","Install the packages using `BiocManager`. If you want to install the library manually, type `BiocManager::install(package_name)` in R."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1193333,"status":"ok","timestamp":1728284836342,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"ZTXvGpVWWAXz","outputId":"180df05b-cc48-4e30-f842-668e3f07ebd8"},"outputs":[{"name":"stderr","output_type":"stream","text":["Installing package into ‘/usr/local/lib/R/site-library’\n","(as ‘lib’ is unspecified)\n","\n","'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n","'help(\"repositories\", package = \"BiocManager\")' for details.\n","Replacement repositories:\n","    CRAN: https://cran.rstudio.com\n","\n","Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)\n","\n","Installing package(s) 'BiocVersion', 'Rsamtools'\n","\n","also installing the dependencies ‘formatR’, ‘lambda.r’, ‘futile.options’, ‘UCSC.utils’, ‘GenomeInfoDbData’, ‘futile.logger’, ‘snow’, ‘BH’, ‘GenomeInfoDb’, ‘GenomicRanges’, ‘Biostrings’, ‘BiocGenerics’, ‘S4Vectors’, ‘IRanges’, ‘XVector’, ‘zlibbioc’, ‘bitops’, ‘BiocParallel’, ‘Rhtslib’\n","\n","\n","Old packages: 'askpass', 'commonmark', 'credentials', 'sys', 'xfun'\n","\n"]},{"name":"stdout","output_type":"stream","text":["Rsamtools is now installed.\n"]},{"name":"stderr","output_type":"stream","text":["'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n","'help(\"repositories\", package = \"BiocManager\")' for details.\n","Replacement repositories:\n","    CRAN: https://cran.rstudio.com\n","\n","Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)\n","\n","Installing package(s) 'GenomicFeatures'\n","\n","also installing the dependencies ‘matrixStats’, ‘abind’, ‘SparseArray’, ‘MatrixGenerics’, ‘S4Arrays’, ‘DelayedArray’, ‘plogr’, ‘png’, ‘SummarizedExperiment’, ‘RCurl’, ‘rjson’, ‘Biobase’, ‘RSQLite’, ‘KEGGREST’, ‘XML’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’, ‘AnnotationDbi’, ‘rtracklayer’\n","\n","\n"]},{"name":"stdout","output_type":"stream","text":["GenomicFeatures is now installed.\n","GenomicAlignments is already installed.\n","BiocParallel is already installed.\n","SummarizedExperiment is already installed.\n"]},{"name":"stderr","output_type":"stream","text":["'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n","'help(\"repositories\", package = \"BiocManager\")' for details.\n","Replacement repositories:\n","    CRAN: https://cran.rstudio.com\n","\n","Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)\n","\n","Installing package(s) 'txdbmaker'\n","\n","also installing the dependencies ‘filelock’, ‘BiocFileCache’, ‘biomaRt’\n","\n","\n"]},{"name":"stdout","output_type":"stream","text":["txdbmaker is now installed.\n"]},{"name":"stderr","output_type":"stream","text":["'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n","'help(\"repositories\", package = \"BiocManager\")' for details.\n","Replacement repositories:\n","    CRAN: https://cran.rstudio.com\n","\n","Bioconductor version 3.19 (BiocManager 1.30.25), R 4.4.1 (2024-06-14)\n","\n","Installing package(s) 'DESeq2'\n","\n","also installing the dependencies ‘locfit’, ‘RcppArmadillo’\n","\n","\n"]},{"name":"stdout","output_type":"stream","text":["DESeq2 is now installed.\n"]}],"source":["packages <- c(\"Rsamtools\", \"GenomicFeatures\", \"GenomicAlignments\", \"BiocParallel\", \"SummarizedExperiment\", \"txdbmaker\", \"DESeq2\")\n","\n","# Install BiocManager\n","if (!requireNamespace(\"BiocManager\", quietly = TRUE)) {\n","  install.packages(\"BiocManager\")\n","}\n","\n","# Install packages\n","for (pkg in packages) {\n","  if (!requireNamespace(pkg, quietly = TRUE)) {\n","    BiocManager::install(pkg, suppressUpdates = TRUE, ask = FALSE)\n","    cat(paste(pkg, \"is now installed.\\n\"))\n","  } else {\n","    cat(paste(pkg, \"is already installed.\\n\"))\n","  }\n","}"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0HqXbj9fm8uW"},"outputs":[],"source":["suppressPackageStartupMessages(library('Rsamtools'))\n","suppressPackageStartupMessages(library('GenomicFeatures'))\n","suppressPackageStartupMessages(library('GenomicAlignments'))\n","suppressPackageStartupMessages(library('BiocParallel'))\n","suppressPackageStartupMessages(library('SummarizedExperiment'))\n","suppressPackageStartupMessages(library('txdbmaker'))\n","suppressPackageStartupMessages(library('DESeq2'))"]},{"cell_type":"markdown","metadata":{"id":"wWDaYJUMm8ua"},"source":["### Set the name of result files"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"o56cI8jGm8ua"},"outputs":[],"source":["TPM_FILE <- 'TPM.csv'\n","TOTAL_DEGs_FILE <- 'Total_DEGs_result.csv'\n","DEGs_FILE <- 'DEGs_result.csv'"]},{"cell_type":"markdown","metadata":{"id":"1TL2-zAJm8ub"},"source":["### Set the experimental informations between control and experimental samples\n","\n","Briefly enter the main keyword of your samples. These keywords will be used for the column names of the DEGs results."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"MnIfkEpdm8ub"},"outputs":[],"source":["CONTROL <- 'WT'   ## e.g. Enter the keyword of control samples\n","EXPERIMENT <- 'K/O_RpoS'   ## e.g. Enter the keyword of experimental samples"]},{"cell_type":"markdown","metadata":{"id":"0oMcSz4KysGy"},"source":["## Step 2. Prepare and upload the files\n","\n","When using Google Colab, as direct mounting of Google Drive is not supported in the R kernel, it is recommended to upload the required files to the Colab environment and configure the working directory for analysis and file management accordingly.\n","\n","### Required files :\n","\n","* Reference genome annotation file (.gff) from **protocol A.1**\n","* Aligned files (.bam) from **protocol B.1**\n","\n","### Set the working directory"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":34},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1728286955619,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"lk5JJjUTY2N4","outputId":"487692bb-be53-4bc2-b520-9bd2fa4552c0"},"outputs":[{"data":{"text/html":["'/content'"],"text/latex":"'/content'","text/markdown":"'/content'","text/plain":["[1] \"/content\""]},"metadata":{},"output_type":"display_data"}],"source":["work_directory <- '/content'  # Enter your working directory.\n","setwd(work_directory)\n","getwd()"]},{"cell_type":"markdown","metadata":{"id":"GvzjlmEGm8uc"},"source":["## Step 3. Prepare Datasheet.csv file\n","\n","This is a comma-separated file created by user according to the following template.\n","Each row contains information of each sample.\n","Information is entered in 2 columns:\n","\n","* **Sample_id** : It is recommended that sample names be designed for easy identification. Replicate names should end with _1, _2, and so on.\n","* **BAM** : Enter the directory of the aligned BAM files."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":260},"executionInfo":{"elapsed":655,"status":"ok","timestamp":1728286965788,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"ziOcH1mMm8uc","outputId":"42b58d1b-17c4-4fca-fd3b-0f5dce542115"},"outputs":[{"name":"stderr","output_type":"stream","text":["Warning message in read.table(file = file, header = header, sep = sep, quote = quote, :\n","“incomplete final line found by readTableHeader on 'RNA-seq_datasheet_example.csv'”\n"]},{"data":{"text/html":["<table class=\"dataframe\">\n","<caption>A data.frame: 4 × 2</caption>\n","<thead>\n","\t<tr><th></th><th scope=col>Sample_id</th><th scope=col>BAM</th></tr>\n","\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n","</thead>\n","<tbody>\n","\t<tr><th scope=row>1</th><td>WT_1      </td><td>mini_RNA-seq_Ecoli_mid-37-1.bam         </td></tr>\n","\t<tr><th scope=row>2</th><td>WT_2      </td><td>mini_RNA-seq_Ecoli_mid-37-2.bam         </td></tr>\n","\t<tr><th scope=row>3</th><td>K/O_RpoS_1</td><td>mini_RNA-seq_Ecoli_del_rpoS_mid-37-1.bam</td></tr>\n","\t<tr><th scope=row>4</th><td>K/O_RpoS_2</td><td>mini_RNA-seq_Ecoli_del_rpoS_mid-37-2.bam</td></tr>\n","</tbody>\n","</table>\n"],"text/latex":"A data.frame: 4 × 2\n\\begin{tabular}{r|ll}\n  & Sample\\_id & BAM\\\\\n  & <chr> & <chr>\\\\\n\\hline\n\t1 & WT\\_1       & mini\\_RNA-seq\\_Ecoli\\_mid-37-1.bam         \\\\\n\t2 & WT\\_2       & mini\\_RNA-seq\\_Ecoli\\_mid-37-2.bam         \\\\\n\t3 & K/O\\_RpoS\\_1 & mini\\_RNA-seq\\_Ecoli\\_del\\_rpoS\\_mid-37-1.bam\\\\\n\t4 & K/O\\_RpoS\\_2 & mini\\_RNA-seq\\_Ecoli\\_del\\_rpoS\\_mid-37-2.bam\\\\\n\\end{tabular}\n","text/markdown":"\nA data.frame: 4 × 2\n\n| <!--/--> | Sample_id &lt;chr&gt; | BAM &lt;chr&gt; |\n|---|---|---|\n| 1 | WT_1       | mini_RNA-seq_Ecoli_mid-37-1.bam          |\n| 2 | WT_2       | mini_RNA-seq_Ecoli_mid-37-2.bam          |\n| 3 | K/O_RpoS_1 | mini_RNA-seq_Ecoli_del_rpoS_mid-37-1.bam |\n| 4 | K/O_RpoS_2 | mini_RNA-seq_Ecoli_del_rpoS_mid-37-2.bam |\n\n","text/plain":["  Sample_id  BAM                                     \n","1 WT_1       mini_RNA-seq_Ecoli_mid-37-1.bam         \n","2 WT_2       mini_RNA-seq_Ecoli_mid-37-2.bam         \n","3 K/O_RpoS_1 mini_RNA-seq_Ecoli_del_rpoS_mid-37-1.bam\n","4 K/O_RpoS_2 mini_RNA-seq_Ecoli_del_rpoS_mid-37-2.bam"]},"metadata":{},"output_type":"display_data"}],"source":["DATASHEET <- 'RNA-seq_datasheet_example.csv'  # Enter your datasheet name.\n","sampleTable <- read.csv(DATASHEET)\n","head(sampleTable)"]},{"cell_type":"markdown","metadata":{"id":"FIpJ7ciSm8ub"},"source":["## Step 4. Load GFF File\n","\n","Set `gff` using the reference genome file from ChEAP.  \n","`makeTxDbFromGFF` loads the GFF file into a database.  \n","`exonsBy` extracts the exons from the GFF file."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1677,"status":"ok","timestamp":1728286961277,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"P0ETTQKBm8ub","outputId":"63de0af6-2c25-4775-ad77-a2250c2728ab"},"outputs":[{"name":"stderr","output_type":"stream","text":["Import genomic features from the file as a GRanges object ... \n","OK\n","\n","Prepare the 'metadata' data frame ... \n","OK\n","\n","Make the TxDb object ... \n","Warning message in .extract_transcripts_from_GRanges(tx_IDX, gr, mcols0$type, mcols0$ID, :\n","“the transcript names (\"tx_name\" column in the TxDb object) imported\n","  from the \"Name\" attribute are not unique”\n","Warning message in makeTxDbFromGRanges(gr, metadata = metadata):\n","“The following transcripts were dropped because their exon ranks could\n","  not be inferred (either because the exons are not on the same\n","  chromosome/strand or because they are not separated by introns): b0149,\n","  b0470, b0484, b1120, b1888, b2592, b3168, b4346, b4795”\n","OK\n","\n"]}],"source":["gff <- 'reference_NC_000913.3.gff'  # Enter the name of your reference genome file.\n","txdb <- makeTxDbFromGFF(gff, format=\"gff\")\n","exons <- exonsBy(txdb, by=\"gene\")"]},{"cell_type":"markdown","metadata":{"id":"gN-gTHcpm8uc"},"source":[" ## Step 5. Load BAM files\n","\n"," Put the name of your bam files into a character vector and check that they all exist on your working directory.\n","\n"," The BamFileList function prepares the BAM files to be processed. The `yieldSize` argument states how many reads can be processed at once (default 2,000,000). This can be increased to speed alignment time, or decreased to reduce memory load."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":72},"executionInfo":{"elapsed":476,"status":"ok","timestamp":1728286986359,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"KlKqsuism8uc","outputId":"bd82a0bc-70e0-4aee-a594-20eafe96e8e3"},"outputs":[{"name":"stdout","output_type":"stream","text":["[1] \"BamFileList has been successfully created.\"\n"]},{"data":{"text/plain":["BamFileList of length 4\n","names(4): mini_RNA-seq_Ecoli_mid-37-1.bam ..."]},"metadata":{},"output_type":"display_data"}],"source":["filenames <- as.character(sampleTable$BAM)\n","\n","if (all(file.exists(filenames))) {\n","  bamfiles <- BamFileList(filenames, yieldSize = 2000000)\n","  print( \"BamFileList has been successfully created.\" )\n","} else {\n","  print( \"Some files do not exist. BamFileList was not created.\" )\n","}\n","\n","bamfiles"]},{"cell_type":"markdown","metadata":{"id":"BOp2e5Xlm8uc"},"source":["## Step 6. Count Reads\n","\n","`summarizeOverlaps` counts the number of reads that overlap each gene in the GFF file. First, we intialize the multiprocessing, using the `workers` argument to set the number of cores to use. The `summarizeOverlaps` arguments are as follows:\n","* `features`: The genomic features loaded in the previous code block\n","* `reads`: The bam files listed above\n","* `mode`: How to deal with potential overlaps. See [HTSeq-count](http://www-huber.embl.de/HTSeq/doc/count.html) documentation.\n","* `singleEnd`: TRUE if single-end, FALSE if paired-end\n","* `ignore.strand`: Whether the strand information is useful for mapping, based on library preparation method\n","    * TRUE: Standard Illumina\n","    * FALSE: Directional Illumina (Ligation), Standard SOLiD, dUTP, NSR, NNSR\n","* `preprocess.reads` (optional): Modify reads before aligning\n","    * invertStrand: Necessary for dUTP, NSR and NNSR library preparation methods\n","* `fragments`: Whether to count unpaired reads"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"KwxPDLk3m8ud"},"outputs":[],"source":["register(MulticoreParam(workers = 4))\n","se <- summarizeOverlaps(features = exons,\n","                        reads = bamfiles,\n","                        mode = \"IntersectionStrict\",\n","                        singleEnd = FALSE,\n","                        ignore.strand = FALSE,\n","                        preprocess.reads = invertStrand,\n","                        fragments = FALSE)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":219},"executionInfo":{"elapsed":488,"status":"ok","timestamp":1728290614222,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"AZbkHhos7SWQ","outputId":"efd44023-976f-4b96-efa3-0b8fe0bcf94c"},"outputs":[{"data":{"text/plain":["class: RangedSummarizedExperiment \n","dim: 4404 4 \n","metadata(0):\n","assays(1): counts\n","rownames(4404): b0001 b0002 ... b4823 b4824\n","rowData names(0):\n","colnames(4): mini_RNA-seq_Ecoli_mid-37-1.bam\n","  mini_RNA-seq_Ecoli_mid-37-2.bam\n","  mini_RNA-seq_Ecoli_del_rpoS_mid-37-1.bam\n","  mini_RNA-seq_Ecoli_del_rpoS_mid-37-2.bam\n","colData names(0):"]},"metadata":{},"output_type":"display_data"}],"source":["se"]},{"cell_type":"markdown","metadata":{"id":"TetUM85Um8ud"},"source":["## Step 7. Generate a dataframe containing raw counts\n","\n","The final counts are stored in the [SummarizedExperiment](https://www.bioconductor.org/help/workflows/rnaseqGene/#summarizedexperiment) object. To view raw counts, use `assay(se)`"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":255},"executionInfo":{"elapsed":505,"status":"ok","timestamp":1728290625976,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"HTquLjL1m8ud","outputId":"533279c4-47df-4702-9627-db13be772e88"},"outputs":[{"data":{"text/html":["<table class=\"dataframe\">\n","<caption>A matrix: 6 × 4 of type int</caption>\n","<thead>\n","\t<tr><th></th><th scope=col>WT_1</th><th scope=col>WT_2</th><th scope=col>K/O_RpoS_1</th><th scope=col>K/O_RpoS_2</th></tr>\n","</thead>\n","<tbody>\n","\t<tr><th scope=row>b0001</th><td>  0</td><td>  0</td><td>  0</td><td>  0</td></tr>\n","\t<tr><th scope=row>b0002</th><td>434</td><td>423</td><td>543</td><td>625</td></tr>\n","\t<tr><th scope=row>b0003</th><td>180</td><td>187</td><td>274</td><td>284</td></tr>\n","\t<tr><th scope=row>b0004</th><td>245</td><td>262</td><td>308</td><td>396</td></tr>\n","\t<tr><th scope=row>b0005</th><td>  2</td><td>  1</td><td>  2</td><td>  6</td></tr>\n","\t<tr><th scope=row>b0006</th><td>  7</td><td> 10</td><td> 10</td><td> 18</td></tr>\n","</tbody>\n","</table>\n"],"text/latex":"A matrix: 6 × 4 of type int\n\\begin{tabular}{r|llll}\n  & WT\\_1 & WT\\_2 & K/O\\_RpoS\\_1 & K/O\\_RpoS\\_2\\\\\n\\hline\n\tb0001 &   0 &   0 &   0 &   0\\\\\n\tb0002 & 434 & 423 & 543 & 625\\\\\n\tb0003 & 180 & 187 & 274 & 284\\\\\n\tb0004 & 245 & 262 & 308 & 396\\\\\n\tb0005 &   2 &   1 &   2 &   6\\\\\n\tb0006 &   7 &  10 &  10 &  18\\\\\n\\end{tabular}\n","text/markdown":"\nA matrix: 6 × 4 of type int\n\n| <!--/--> | WT_1 | WT_2 | K/O_RpoS_1 | K/O_RpoS_2 |\n|---|---|---|---|---|\n| b0001 |   0 |   0 |   0 |   0 |\n| b0002 | 434 | 423 | 543 | 625 |\n| b0003 | 180 | 187 | 274 | 284 |\n| b0004 | 245 | 262 | 308 | 396 |\n| b0005 |   2 |   1 |   2 |   6 |\n| b0006 |   7 |  10 |  10 |  18 |\n\n","text/plain":["      WT_1 WT_2 K/O_RpoS_1 K/O_RpoS_2\n","b0001   0    0    0          0       \n","b0002 434  423  543        625       \n","b0003 180  187  274        284       \n","b0004 245  262  308        396       \n","b0005   2    1    2          6       \n","b0006   7   10   10         18       "]},"metadata":{},"output_type":"display_data"}],"source":["metadata <- sampleTable\n","colData(se) <- DataFrame(metadata)\n","colnames(se) <- colData(se)$Sample_id\n","\n","head(assay(se))"]},{"cell_type":"markdown","metadata":{"id":"FfFzB-qMm8ud"},"source":["To assign groups for comparison, remove the replicate subscript from the colnames."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"9ZdP9ZUxm8ud"},"outputs":[],"source":["colData(se)$group <- gsub('.{2}$', '', colnames(se))"]},{"cell_type":"markdown","metadata":{"id":"yJY9PDJJm8ud"},"source":["## Step 8. Convert the dataframe into DESeqDataSet (DDS)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":418},"executionInfo":{"elapsed":467,"status":"ok","timestamp":1728290634901,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"9F6RI9XHm8ud","outputId":"202a78c9-2069-41dc-c9d0-ca52a3581a1a"},"outputs":[{"name":"stderr","output_type":"stream","text":["Warning message in DESeqDataSet(se, design = ~group):\n","“some variables in design formula are characters, converting to factors”\n","  Note: levels of factors in the design contain characters other than\n","  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n","  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n","  for column names in R. [This is a message, not a warning or an error]\n","\n"]},{"data":{"text/html":["4404"],"text/latex":"4404","text/markdown":"4404","text/plain":["[1] 4404"]},"metadata":{},"output_type":"display_data"},{"data":{"text/html":["3819"],"text/latex":"3819","text/markdown":"3819","text/plain":["[1] 3819"]},"metadata":{},"output_type":"display_data"},{"data":{"text/html":["<table class=\"dataframe\">\n","<caption>A matrix: 6 × 4 of type int</caption>\n","<thead>\n","\t<tr><th></th><th scope=col>WT_1</th><th scope=col>WT_2</th><th scope=col>K/O_RpoS_1</th><th scope=col>K/O_RpoS_2</th></tr>\n","</thead>\n","<tbody>\n","\t<tr><th scope=row>b0002</th><td>434</td><td>423</td><td>543</td><td>625</td></tr>\n","\t<tr><th scope=row>b0003</th><td>180</td><td>187</td><td>274</td><td>284</td></tr>\n","\t<tr><th scope=row>b0004</th><td>245</td><td>262</td><td>308</td><td>396</td></tr>\n","\t<tr><th scope=row>b0005</th><td>  2</td><td>  1</td><td>  2</td><td>  6</td></tr>\n","\t<tr><th scope=row>b0006</th><td>  7</td><td> 10</td><td> 10</td><td> 18</td></tr>\n","\t<tr><th scope=row>b0007</th><td>  2</td><td>  1</td><td>  3</td><td>  5</td></tr>\n","</tbody>\n","</table>\n"],"text/latex":"A matrix: 6 × 4 of type int\n\\begin{tabular}{r|llll}\n  & WT\\_1 & WT\\_2 & K/O\\_RpoS\\_1 & K/O\\_RpoS\\_2\\\\\n\\hline\n\tb0002 & 434 & 423 & 543 & 625\\\\\n\tb0003 & 180 & 187 & 274 & 284\\\\\n\tb0004 & 245 & 262 & 308 & 396\\\\\n\tb0005 &   2 &   1 &   2 &   6\\\\\n\tb0006 &   7 &  10 &  10 &  18\\\\\n\tb0007 &   2 &   1 &   3 &   5\\\\\n\\end{tabular}\n","text/markdown":"\nA matrix: 6 × 4 of type int\n\n| <!--/--> | WT_1 | WT_2 | K/O_RpoS_1 | K/O_RpoS_2 |\n|---|---|---|---|---|\n| b0002 | 434 | 423 | 543 | 625 |\n| b0003 | 180 | 187 | 274 | 284 |\n| b0004 | 245 | 262 | 308 | 396 |\n| b0005 |   2 |   1 |   2 |   6 |\n| b0006 |   7 |  10 |  10 |  18 |\n| b0007 |   2 |   1 |   3 |   5 |\n\n","text/plain":["      WT_1 WT_2 K/O_RpoS_1 K/O_RpoS_2\n","b0002 434  423  543        625       \n","b0003 180  187  274        284       \n","b0004 245  262  308        396       \n","b0005   2    1    2          6       \n","b0006   7   10   10         18       \n","b0007   2    1    3          5       "]},"metadata":{},"output_type":"display_data"}],"source":["dds <- DESeqDataSet(se, design = ~group)\n","nrow(dds)\n","dds <- dds[rowSums(assay(dds)) > 0, ]\n","nrow(dds)\n","\n","head(assay(dds))"]},{"cell_type":"markdown","metadata":{"id":"vxNLZbtPm8ue"},"source":["## Step 9. Calculate FPKM\n","\n","Calculate the fragments per kilobase of transcript per million mapped reads (FPKM) value from the DDS object."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":347},"executionInfo":{"elapsed":470,"status":"ok","timestamp":1728290642766,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"gI-qym9pm8ue","outputId":"e40021e7-54a8-4af5-dc52-d61c151233db"},"outputs":[{"name":"stderr","output_type":"stream","text":["  Note: levels of factors in the design contain characters other than\n","  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n","  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n","  for column names in R. [This is a message, not a warning or an error]\n","\n"]},{"data":{"text/html":["<table class=\"dataframe\">\n","<caption>A matrix: 6 × 4 of type dbl</caption>\n","<thead>\n","\t<tr><th></th><th scope=col>WT_1</th><th scope=col>WT_2</th><th scope=col>K/O_RpoS_1</th><th scope=col>K/O_RpoS_2</th></tr>\n","</thead>\n","<tbody>\n","\t<tr><th scope=row>b0002</th><td>1411.29216</td><td>1357.818042</td><td>1532.89762</td><td>1423.26415</td></tr>\n","\t<tr><th scope=row>b0003</th><td>1544.16439</td><td>1583.567724</td><td>2040.59920</td><td>1706.15177</td></tr>\n","\t<tr><th scope=row>b0004</th><td>1524.11636</td><td>1608.893657</td><td>1663.37005</td><td>1725.14466</td></tr>\n","\t<tr><th scope=row>b0005</th><td>  53.77515</td><td>  26.541511</td><td>  46.68397</td><td> 112.97470</td></tr>\n","\t<tr><th scope=row>b0006</th><td>  72.09188</td><td> 101.662858</td><td>  89.40760</td><td> 129.81925</td></tr>\n","\t<tr><th scope=row>b0007</th><td>  11.19064</td><td>   5.523303</td><td>  14.57244</td><td>  19.59175</td></tr>\n","</tbody>\n","</table>\n"],"text/latex":"A matrix: 6 × 4 of type dbl\n\\begin{tabular}{r|llll}\n  & WT\\_1 & WT\\_2 & K/O\\_RpoS\\_1 & K/O\\_RpoS\\_2\\\\\n\\hline\n\tb0002 & 1411.29216 & 1357.818042 & 1532.89762 & 1423.26415\\\\\n\tb0003 & 1544.16439 & 1583.567724 & 2040.59920 & 1706.15177\\\\\n\tb0004 & 1524.11636 & 1608.893657 & 1663.37005 & 1725.14466\\\\\n\tb0005 &   53.77515 &   26.541511 &   46.68397 &  112.97470\\\\\n\tb0006 &   72.09188 &  101.662858 &   89.40760 &  129.81925\\\\\n\tb0007 &   11.19064 &    5.523303 &   14.57244 &   19.59175\\\\\n\\end{tabular}\n","text/markdown":"\nA matrix: 6 × 4 of type dbl\n\n| <!--/--> | WT_1 | WT_2 | K/O_RpoS_1 | K/O_RpoS_2 |\n|---|---|---|---|---|\n| b0002 | 1411.29216 | 1357.818042 | 1532.89762 | 1423.26415 |\n| b0003 | 1544.16439 | 1583.567724 | 2040.59920 | 1706.15177 |\n| b0004 | 1524.11636 | 1608.893657 | 1663.37005 | 1725.14466 |\n| b0005 |   53.77515 |   26.541511 |   46.68397 |  112.97470 |\n| b0006 |   72.09188 |  101.662858 |   89.40760 |  129.81925 |\n| b0007 |   11.19064 |    5.523303 |   14.57244 |   19.59175 |\n\n","text/plain":["      WT_1       WT_2        K/O_RpoS_1 K/O_RpoS_2\n","b0002 1411.29216 1357.818042 1532.89762 1423.26415\n","b0003 1544.16439 1583.567724 2040.59920 1706.15177\n","b0004 1524.11636 1608.893657 1663.37005 1725.14466\n","b0005   53.77515   26.541511   46.68397  112.97470\n","b0006   72.09188  101.662858   89.40760  129.81925\n","b0007   11.19064    5.523303   14.57244   19.59175"]},"metadata":{},"output_type":"display_data"}],"source":["fpkm_data <- fpkm(dds)\n","\n","head(fpkm_data)"]},{"cell_type":"markdown","metadata":{"id":"H9Dy4brRm8ue"},"source":["## Step 10. Calculate TPM\n","\n","Convert the calculated FPKM values to transcripts per million (TPM) values."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":255},"executionInfo":{"elapsed":478,"status":"ok","timestamp":1728290649165,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"O11jVouxm8ue","outputId":"bede2c4d-ab97-4138-856c-f77cbc9cf1d9"},"outputs":[{"data":{"text/html":["<table class=\"dataframe\">\n","<caption>A matrix: 6 × 4 of type dbl</caption>\n","<thead>\n","\t<tr><th></th><th scope=col>WT_1</th><th scope=col>WT_2</th><th scope=col>K/O_RpoS_1</th><th scope=col>K/O_RpoS_2</th></tr>\n","</thead>\n","<tbody>\n","\t<tr><th scope=row>b0002</th><td>1273.02226</td><td>1256.404386</td><td>1362.83665</td><td>1277.47400</td></tr>\n","\t<tr><th scope=row>b0003</th><td>1392.87647</td><td>1465.293119</td><td>1814.21338</td><td>1531.38441</td></tr>\n","\t<tr><th scope=row>b0004</th><td>1374.79262</td><td>1488.727491</td><td>1478.83436</td><td>1548.43179</td></tr>\n","\t<tr><th scope=row>b0005</th><td>  48.50659</td><td>  24.559160</td><td>  41.50481</td><td> 101.40229</td></tr>\n","\t<tr><th scope=row>b0006</th><td>  65.02875</td><td>  94.069792</td><td>  79.48864</td><td> 116.52139</td></tr>\n","\t<tr><th scope=row>b0007</th><td>  10.09425</td><td>   5.110775</td><td>  12.95576</td><td>  17.58489</td></tr>\n","</tbody>\n","</table>\n"],"text/latex":"A matrix: 6 × 4 of type dbl\n\\begin{tabular}{r|llll}\n  & WT\\_1 & WT\\_2 & K/O\\_RpoS\\_1 & K/O\\_RpoS\\_2\\\\\n\\hline\n\tb0002 & 1273.02226 & 1256.404386 & 1362.83665 & 1277.47400\\\\\n\tb0003 & 1392.87647 & 1465.293119 & 1814.21338 & 1531.38441\\\\\n\tb0004 & 1374.79262 & 1488.727491 & 1478.83436 & 1548.43179\\\\\n\tb0005 &   48.50659 &   24.559160 &   41.50481 &  101.40229\\\\\n\tb0006 &   65.02875 &   94.069792 &   79.48864 &  116.52139\\\\\n\tb0007 &   10.09425 &    5.110775 &   12.95576 &   17.58489\\\\\n\\end{tabular}\n","text/markdown":"\nA matrix: 6 × 4 of type dbl\n\n| <!--/--> | WT_1 | WT_2 | K/O_RpoS_1 | K/O_RpoS_2 |\n|---|---|---|---|---|\n| b0002 | 1273.02226 | 1256.404386 | 1362.83665 | 1277.47400 |\n| b0003 | 1392.87647 | 1465.293119 | 1814.21338 | 1531.38441 |\n| b0004 | 1374.79262 | 1488.727491 | 1478.83436 | 1548.43179 |\n| b0005 |   48.50659 |   24.559160 |   41.50481 |  101.40229 |\n| b0006 |   65.02875 |   94.069792 |   79.48864 |  116.52139 |\n| b0007 |   10.09425 |    5.110775 |   12.95576 |   17.58489 |\n\n","text/plain":["      WT_1       WT_2        K/O_RpoS_1 K/O_RpoS_2\n","b0002 1273.02226 1256.404386 1362.83665 1277.47400\n","b0003 1392.87647 1465.293119 1814.21338 1531.38441\n","b0004 1374.79262 1488.727491 1478.83436 1548.43179\n","b0005   48.50659   24.559160   41.50481  101.40229\n","b0006   65.02875   94.069792   79.48864  116.52139\n","b0007   10.09425    5.110775   12.95576   17.58489"]},"metadata":{},"output_type":"display_data"}],"source":["tpm_data <- sweep(fpkm_data,2,colSums(fpkm_data),`/`)*1e6  # Divide each column by the column sum (times 1e6)\n","head(tpm_data)\n","\n","write.csv(tpm_data, file = TPM_FILE)"]},{"cell_type":"markdown","metadata":{"id":"EsAoTzmmm8ue"},"source":["## Step 11. Find DEGs\n","\n","Calculate the DEGs from the DDS object using the ‘DESeq’ function provided by the DESeq2 package.\n","\n","Each column has the following meaning:\n","* **baseMean**: The average normalized expression of the gene across both experimental and control groups.\n","* **log2FoldChange**: The log2 difference in gene expression between the experimental and control groups. Positive means higher expression in the experimental group, negative means higher in the control.\n","* **lfcSE**: The standard error of the log2FoldChange.\n","* **stat**: The Wald test statistic indicating the significance of gene expression change.\n","* **pvalue**: The probability that the gene expression difference occurred by chance.\n","* **padj**: The adjusted p-value, correcting for multiple testing (False Discovery Rate; FDR). It indicates the statistical significance of the expression difference, with smaller values showing more confidence in the result."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":690},"executionInfo":{"elapsed":4126,"status":"ok","timestamp":1728290659571,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"},"user_tz":-540},"id":"1V6DMPO6m8ue","outputId":"ef47ff28-4fe5-4a9a-d446-15783750bd56"},"outputs":[{"name":"stderr","output_type":"stream","text":["estimating size factors\n","\n","  Note: levels of factors in the design contain characters other than\n","  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n","  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n","  for column names in R. [This is a message, not a warning or an error]\n","\n","estimating dispersions\n","\n","gene-wise dispersion estimates\n","\n","mean-dispersion relationship\n","\n","  Note: levels of factors in the design contain characters other than\n","  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n","  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n","  for column names in R. [This is a message, not a warning or an error]\n","\n","final dispersion estimates\n","\n","fitting model and testing\n","\n"]},{"data":{"text/html":["<table class=\"dataframe\">\n","<caption>A data.frame: 6 × 6</caption>\n","<thead>\n","\t<tr><th></th><th scope=col>baseMean</th><th scope=col>log2FoldChange</th><th scope=col>lfcSE</th><th scope=col>stat</th><th scope=col>pvalue</th><th scope=col>padj</th></tr>\n","\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n","</thead>\n","<tbody>\n","\t<tr><th scope=row>b0002</th><td>495.272229</td><td>0.09300726</td><td>0.1260783</td><td>0.7376943</td><td>0.4607002</td><td>0.9995757</td></tr>\n","\t<tr><th scope=row>b0003</th><td>225.420845</td><td>0.25745488</td><td>0.1897037</td><td>1.3571418</td><td>0.1747362</td><td>0.9995757</td></tr>\n","\t<tr><th scope=row>b0004</th><td>294.898225</td><td>0.11370526</td><td>0.1610864</td><td>0.7058649</td><td>0.4802721</td><td>0.9995757</td></tr>\n","\t<tr><th scope=row>b0005</th><td>  2.510672</td><td>1.00550820</td><td>1.7566098</td><td>0.5724141</td><td>0.5670415</td><td>       NA</td></tr>\n","\t<tr><th scope=row>b0006</th><td> 10.733935</td><td>0.34137971</td><td>0.8428585</td><td>0.4050261</td><td>0.6854583</td><td>0.9995757</td></tr>\n","\t<tr><th scope=row>b0007</th><td>  2.557887</td><td>1.03742772</td><td>1.7280046</td><td>0.6003617</td><td>0.5482652</td><td>       NA</td></tr>\n","</tbody>\n","</table>\n"],"text/latex":"A data.frame: 6 × 6\n\\begin{tabular}{r|llllll}\n  & baseMean & log2FoldChange & lfcSE & stat & pvalue & padj\\\\\n  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n\\hline\n\tb0002 & 495.272229 & 0.09300726 & 0.1260783 & 0.7376943 & 0.4607002 & 0.9995757\\\\\n\tb0003 & 225.420845 & 0.25745488 & 0.1897037 & 1.3571418 & 0.1747362 & 0.9995757\\\\\n\tb0004 & 294.898225 & 0.11370526 & 0.1610864 & 0.7058649 & 0.4802721 & 0.9995757\\\\\n\tb0005 &   2.510672 & 1.00550820 & 1.7566098 & 0.5724141 & 0.5670415 &        NA\\\\\n\tb0006 &  10.733935 & 0.34137971 & 0.8428585 & 0.4050261 & 0.6854583 & 0.9995757\\\\\n\tb0007 &   2.557887 & 1.03742772 & 1.7280046 & 0.6003617 & 0.5482652 &        NA\\\\\n\\end{tabular}\n","text/markdown":"\nA data.frame: 6 × 6\n\n| <!--/--> | baseMean &lt;dbl&gt; | log2FoldChange &lt;dbl&gt; | lfcSE &lt;dbl&gt; | stat &lt;dbl&gt; | pvalue &lt;dbl&gt; | padj &lt;dbl&gt; |\n|---|---|---|---|---|---|---|\n| b0002 | 495.272229 | 0.09300726 | 0.1260783 | 0.7376943 | 0.4607002 | 0.9995757 |\n| b0003 | 225.420845 | 0.25745488 | 0.1897037 | 1.3571418 | 0.1747362 | 0.9995757 |\n| b0004 | 294.898225 | 0.11370526 | 0.1610864 | 0.7058649 | 0.4802721 | 0.9995757 |\n| b0005 |   2.510672 | 1.00550820 | 1.7566098 | 0.5724141 | 0.5670415 |        NA |\n| b0006 |  10.733935 | 0.34137971 | 0.8428585 | 0.4050261 | 0.6854583 | 0.9995757 |\n| b0007 |   2.557887 | 1.03742772 | 1.7280046 | 0.6003617 | 0.5482652 |        NA |\n\n","text/plain":["      baseMean   log2FoldChange lfcSE     stat      pvalue    padj     \n","b0002 495.272229 0.09300726     0.1260783 0.7376943 0.4607002 0.9995757\n","b0003 225.420845 0.25745488     0.1897037 1.3571418 0.1747362 0.9995757\n","b0004 294.898225 0.11370526     0.1610864 0.7058649 0.4802721 0.9995757\n","b0005   2.510672 1.00550820     1.7566098 0.5724141 0.5670415        NA\n","b0006  10.733935 0.34137971     0.8428585 0.4050261 0.6854583 0.9995757\n","b0007   2.557887 1.03742772     1.7280046 0.6003617 0.5482652        NA"]},"metadata":{},"output_type":"display_data"}],"source":["dds <- DESeq( dds )\n","res <- results(dds,contrast = c( 'group', EXPERIMENT , CONTROL ))\n","resTable <- data.frame( res@listData, row.names = res@rownames )\n","head( resTable )\n","\n","write.csv(resTable, TOTAL_DEGs_FILE, row.names=T, quote=T)"]},{"cell_type":"markdown","metadata":{"id":"dOGdsUGlm8uf"},"source":["## Step 12. Identify the DEGs using cut-offs\n","\n","To identify DEGs that are particularly affected by the experimental conditions, a cut-off is set to output only those DEGs that meet the conditions."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"QX2vlWzkm8uf"},"outputs":[],"source":["deg <- resTable[complete.cases(resTable[,c('padj','log2FoldChange')]) &\n","                            resTable$padj < 0.05 &\n","                            abs(resTable$log2FoldChange) > 1.0,]\n","\n","write.csv(deg, DEGs_FILE, row.names=T, quote=T)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"liQJOA2hJrrQ"},"outputs":[],"source":[]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"R","name":"ir"},"language_info":{"name":"R"},"toc":{"nav_menu":{},"number_sections":true,"sideBar":true,"skip_h1_title":false,"title_cell":"Table of Contents","title_sidebar":"Contents","toc_cell":true,"toc_position":{},"toc_section_display":true,"toc_window_display":false}},"nbformat":4,"nbformat_minor":0}