{"cells":[{"cell_type":"markdown","id":"52d949fd","metadata":{"id":"52d949fd"},"source":["# Python-based automated pipeline for classifying the regulatory roles of each curated peak from ChIP-exo datasets"]},{"cell_type":"markdown","id":"9fedd9fe","metadata":{"id":"9fedd9fe"},"source":["## Protocol A.4 - Classification of Regulatory Roles in Each Curated Peak"]},{"cell_type":"markdown","id":"4ae969f3","metadata":{"id":"4ae969f3"},"source":["## Step 1. Set-up\n","\n","### Install and import libraries\n","Install the packages using `pip`.\n","If you want to install the library manually, type `pip3 install package_name` in terminal."]},{"cell_type":"code","execution_count":null,"id":"922d1536","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"922d1536","executionInfo":{"status":"ok","timestamp":1728282607290,"user_tz":-540,"elapsed":399,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"29b34b76-2d8e-4b3d-f4b8-9d9ca30514e8"},"outputs":[{"output_type":"stream","name":"stdout","text":["pandas is installed.\n"]}],"source":["import os, importlib, re\n","\n","libraries = { 'pandas': 'pd' }\n","\n","for lib, package_name in libraries.items():\n","    try:\n","        importlib.import_module(lib)\n","        print(f\"{lib} is installed.\")\n","    except ImportError:\n","        print(f\"{lib} is not installed. Installing {lib}...\")\n","        os.system(f'pip3 install {package_name}')"]},{"cell_type":"code","source":["import pandas as pd"],"metadata":{"id":"vldL4g501mi9"},"id":"vldL4g501mi9","execution_count":null,"outputs":[]},{"cell_type":"markdown","id":"6a45cb88","metadata":{"id":"6a45cb88"},"source":["### Function\n","Please run the cell without modification"]},{"cell_type":"code","execution_count":null,"id":"c5381cfb","metadata":{"id":"c5381cfb"},"outputs":[],"source":["def extract_locus_tag_number(locus_tag):\n","    match = re.search(r'\\d+', locus_tag)\n","    if match:\n","        return int(match.group())\n","    else:\n","        return None\n","\n","def read_chip_exo_peaks(peak_file):\n","    print(\"Reading ChIP-exo peak calling data...\")\n","    peaks = pd.read_csv(peak_file, sep='\\t', header=None, names=[\n","        'chrom', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute'])\n","    peaks['midpoint'] = ((peaks['start'] + peaks['end']) / 2).round().astype(int)\n","    peaks['Peak_name'] = peaks['attribute'].apply(lambda x: x.strip('\"').split(',')[0])\n","    print(\"ChIP-exo peak data successfully read.\")\n","    return peaks\n","\n","def extract_locus_tags(gff_file):\n","    locus_tags = []\n","    try:\n","        print(\"Reading reference genome annotation data...\")\n","        with open(gff_file, 'r') as file:\n","            for line in file:\n","                if line.startswith('#'):\n","                    continue\n","                parts = line.strip().split('\\t')\n","                if len(parts) < 9:\n","                    continue\n","                attributes_str = parts[8]\n","                attributes = extract_attributes(attributes_str)\n","                locus_tag = attributes.get('locus_tag', '')\n","                gene_type = attributes.get('type','')\n","                if locus_tag:\n","                    locus_tags.append({\n","                        'chrom': parts[0],\n","                        'start': int(parts[3]),\n","                        'end': int(parts[4]),\n","                        'strand': parts[6],\n","                        'locus_tag': locus_tag,\n","                        'gene_type': gene_type\n","                    })\n","        print(\"Reference genome annotation data successfully read.\")\n","    except Exception as e:\n","        print(f\"Error reading GFF file: {e}\")\n","    return pd.DataFrame(locus_tags)\n","\n","def extract_attributes(attributes_str):\n","    attributes = {}\n","    try:\n","        for attribute in attributes_str.split(';'):\n","            if '=' in attribute:\n","                key, value = attribute.split('=')\n","                attributes[key] = value\n","    except Exception as e:\n","        print(f\"Error extracting attributes: {e}\")\n","    return attributes\n","\n","def read_transcription_units(tu_file):\n","    try:\n","        print(\"Reading transcription unit data...\")\n","        transcription_units = pd.read_csv(tu_file, sep='\\t', header=None, names=['TU'])\n","        transcription_units['TU'] = transcription_units['TU'].apply(sort_tu)\n","        print(\"Transcription unit data successfully read and sorted.\")\n","        return transcription_units\n","    except Exception as e:\n","        print(f\"Error reading TU file: {e}\")\n","        return pd.DataFrame()\n","\n","def sort_tu(tu):\n","    try:\n","        tu_list = tu.split(',')\n","        tu_list_sorted = sorted(tu_list)\n","        return ','.join(tu_list_sorted)\n","    except Exception as e:\n","        print(f\"Error sorting TU: {e}\")\n","        return tu\n","\n","def select_largest_tu(regulatory_genes, transcription_units):\n","    largest_tu = ''\n","    for tu in transcription_units['TU']:\n","        tu_list = tu.split(',')\n","        if any(gene in tu_list for gene in regulatory_genes):\n","            current_tu = ','.join(tu_list)\n","            if len(current_tu.split(',')) > len(largest_tu.split(',')):\n","                largest_tu = current_tu\n","    return largest_tu\n","\n","def estimate_promoter(annotation, gene_to_promoter =300):\n","    if annotation['strand'] == '+':\n","        promoter_start = max(0, annotation['start'] - gene_to_promoter )\n","        promoter_end = annotation['start']\n","    else:\n","        promoter_start = annotation['end']\n","        promoter_end = annotation['end'] + gene_to_promoter\n","    return promoter_start, promoter_end\n","\n","def calculate_real_tu_by_position(regulatory_genes, tu, strand, regulatory_positions, annotations, filter_cds=True):\n","    tu_list = tu.split(',')\n","    real_tu = []\n","    regulatory_positions_list = [regulatory_positions.get(gene) for gene in regulatory_genes if regulatory_positions.get(gene) is not None]\n","    if not regulatory_positions_list:\n","        return ''\n","    if filter_cds:\n","        tu_list = [gene for gene in tu_list if not annotations.loc[annotations['locus_tag'] == gene, 'gene_type'].empty and\n","                   annotations.loc[annotations['locus_tag'] == gene, 'gene_type'].values[0] == 'CDS']\n","    if strand == '+':\n","        min_reg_position = min(regulatory_positions_list)\n","        real_tu = [gene for gene in tu_list if regulatory_positions.get(gene) and regulatory_positions[gene] >= min_reg_position]\n","    elif strand == '-':\n","        max_reg_position = max(regulatory_positions_list)\n","        real_tu = [gene for gene in tu_list if regulatory_positions.get(gene) and regulatory_positions[gene] <= max_reg_position]\n","    return ','.join(real_tu)\n","\n","def find_regulatory_roles(peaks, annotations, transcription_units=None, gene_to_promoter =300, filter_cds=True):\n","    print(\"Starting regulatory role analysis...\")\n","    regulatory_roles = []\n","    use_tu = transcription_units is not None\n","    regulatory_positions = annotations.set_index('locus_tag')[['start', 'end']].mean(axis=1).to_dict()\n","    for index, peak in peaks.iterrows():\n","        if index % 100 == 0:\n","            print(f\"Processing peak {index + 1}/{len(peaks)}...\")\n","        matching_annotations = annotations[annotations['chrom'] == peak['chrom']]\n","        matching_annotations['promoter_start'], matching_annotations['promoter_end'] = zip(\n","            *matching_annotations.apply(lambda row: estimate_promoter(row, gene_to_promoter ), axis=1)\n","        )\n","        regulatory_genes_plus = matching_annotations[\n","            (matching_annotations['promoter_start'] <= peak['midpoint']) &\n","            (matching_annotations['promoter_end'] >= peak['midpoint']) &\n","            (matching_annotations['strand'] == '+')\n","        ]['locus_tag'].tolist()\n","        regulatory_genes_minus = matching_annotations[\n","            (matching_annotations['promoter_start'] <= peak['midpoint']) &\n","            (matching_annotations['promoter_end'] >= peak['midpoint']) &\n","            (matching_annotations['strand'] == '-')\n","        ]['locus_tag'].tolist()\n","\n","        tu_plus = ''\n","        real_tu_plus = ''\n","        if use_tu:\n","            tu_plus = select_largest_tu(regulatory_genes_plus, transcription_units)\n","            real_tu_plus = calculate_real_tu_by_position(\n","                regulatory_genes_plus, tu_plus, '+', regulatory_positions, annotations, filter_cds) if tu_plus else ''\n","        else:\n","            tu_plus = ','.join(regulatory_genes_plus)\n","            real_tu_plus = calculate_real_tu_by_position(\n","                regulatory_genes_plus, tu_plus, '+', regulatory_positions, annotations, filter_cds) if tu_plus else ''\n","\n","        tu_minus = ''\n","        real_tu_minus = ''\n","        if use_tu:\n","            tu_minus = select_largest_tu(regulatory_genes_minus, transcription_units)\n","            real_tu_minus = calculate_real_tu_by_position(\n","                regulatory_genes_minus, tu_minus, '-', regulatory_positions, annotations, filter_cds) if tu_minus else ''\n","        else:\n","            tu_minus = ','.join(regulatory_genes_minus)\n","            real_tu_minus = calculate_real_tu_by_position(\n","                regulatory_genes_minus, tu_minus, '-', regulatory_positions, annotations, filter_cds) if tu_minus else ''\n","\n","        regulatory_roles.append({\n","            'Start': peak['start'],\n","            'End': peak['end'],\n","            'Peak_name': peak['Peak_name'],\n","            'Regulatory': '0' if regulatory_genes_plus or regulatory_genes_minus else 'x',\n","            'Bidirection': '0' if regulatory_genes_plus and regulatory_genes_minus else 'x',\n","            'TU_plus_strand': tu_plus,\n","            'Real_TU_plus_strand': real_tu_plus,\n","            'TU_minus_strand': tu_minus,\n","            'Real_TU_minus_strand': real_tu_minus\n","        })\n","    print(\"Regulatory role analysis complete.\")\n","    return pd.DataFrame(regulatory_roles)\n","\n","def save_regulatory_roles(regulatory_roles, output_file):\n","    try:\n","        print(f\"Saving results to {output_file}...\")\n","        regulatory_roles.to_csv(output_file, sep='\\t', index=False)\n","        print(f\"Results successfully saved to {output_file}.\")\n","    except Exception as e:\n","        print(f\"Error saving regulatory roles: {e}\")\n","\n","def Defining_regulatory_roles_of_peaks(working_directroy, peak_file, gff_file, output_file, tu_file=None, gene_to_promoter =300, filter_cds=True):\n","    os.chdir(working_directory)\n","    print(f'Current working directiory : {os.getcwd()}')\n","\n","    peaks = read_chip_exo_peaks(peak_file)\n","    annotations = extract_locus_tags(gff_file)\n","    transcription_units = read_transcription_units(tu_file) if tu_file else None\n","    regulatory_roles = find_regulatory_roles(peaks, annotations, transcription_units, gene_to_promoter , filter_cds)\n","    save_regulatory_roles(regulatory_roles, output_file)\n","    print(\"Processing completed\")"]},{"cell_type":"markdown","id":"a00ddc56","metadata":{"id":"a00ddc56"},"source":["### (Optional) Google Drive Mount\n","\n","If you are using Google Colaboratory, mount Google Drive to access files stored in your Google Drive."]},{"cell_type":"code","execution_count":null,"id":"4ce38c3c","metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"4ce38c3c","executionInfo":{"status":"ok","timestamp":1728282623724,"user_tz":-540,"elapsed":3044,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"7bd51329-cf97-437c-840a-527d255a510c"},"outputs":[{"output_type":"stream","name":"stdout","text":["Drive already mounted at /content/mnt; to attempt to forcibly remount, call drive.mount(\"/content/mnt\", force_remount=True).\n"]}],"source":["from google.colab import drive, files\n","\n","drive.mount('/content/mnt')"]},{"cell_type":"markdown","id":"2965668c","metadata":{"id":"2965668c"},"source":["## Step 2. Determine the appropriate distance range between gene and promoter\n","\n","Set the appropriate distance range between gene and promoter in interested strain. If the value is hard to determine, then try different values and compare with them."]},{"cell_type":"code","execution_count":null,"id":"81217c3e","metadata":{"id":"81217c3e"},"outputs":[],"source":["gene_to_promoter = 300  ## Enter the value (Default : 300)"]},{"cell_type":"markdown","id":"05ba82ba","metadata":{"id":"05ba82ba"},"source":["## Step 3. Run the automated pipeline\n","\n","### Define the regulatory role and target genes of peaks\n","\n","Enter the information of files.\n","* **working_directory** : Enter your working directory.\n","* **peak_file** : Enter the name of visualized cruated peak file (.gff) from Protocol A.3.\n","* **gff_file** : Enter the name of visualized reference genome file (.gff) from ChEAP.\n","* **output_file** : Enter the name of result file.\n","* **tu_file** : (optional) Enter the directory of the file (.tsv) containing the transcription units. If you don't have the TU file, then set to `None`. (Default : `None`)\n","* **filter_cds** : Set to `True` if you want to include only CDS genes in the output file. (Default : `True`)"]},{"cell_type":"code","execution_count":null,"id":"0f606f0a","metadata":{"id":"0f606f0a"},"outputs":[],"source":["working_directory =  '/content/mnt/My Drive/Test/ChIP-exo' ##'/home/user/example'\n","peak_file = 'Curated_peak.gff'\n","gff_file = 'reference_NC_000913.3.gff'\n","output_file = 'output.tsv'\n","tu_file = 'MG1655_tu_info.tsv'\n","filter_cds = True"]},{"cell_type":"code","execution_count":null,"id":"d84597e5","metadata":{"id":"d84597e5","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1728282811402,"user_tz":-540,"elapsed":124884,"user":{"displayName":"Minchang Jang","userId":"05813724370473662280"}},"outputId":"74cc02fc-7d00-485e-99a0-fcb1d1d3f463"},"outputs":[{"output_type":"stream","name":"stdout","text":["Current working directiory : /content/mnt/My Drive/Test/ChIP-exo\n","Reading ChIP-exo peak calling data...\n","ChIP-exo peak data successfully read.\n","Reading reference genome annotation data...\n","Reference genome annotation data successfully read.\n","Reading transcription unit data...\n","Transcription unit data successfully read and sorted.\n","Starting regulatory role analysis...\n","Processing peak 1/1566...\n","Processing peak 101/1566...\n","Processing peak 201/1566...\n","Processing peak 301/1566...\n","Processing peak 401/1566...\n","Processing peak 501/1566...\n","Processing peak 601/1566...\n","Processing peak 701/1566...\n","Processing peak 801/1566...\n","Processing peak 901/1566...\n","Processing peak 1001/1566...\n","Processing peak 1101/1566...\n","Processing peak 1201/1566...\n","Processing peak 1301/1566...\n","Processing peak 1401/1566...\n","Processing peak 1501/1566...\n","Regulatory role analysis complete.\n","Saving results to output.tsv...\n","Results successfully saved to output.tsv.\n","Processing completed\n"]}],"source":["Defining_regulatory_roles_of_peaks( working_directory, peak_file , gff_file, output_file, tu_file, gene_to_promoter , filter_cds )"]},{"cell_type":"code","source":[],"metadata":{"id":"D0i9QTQbuqps"},"id":"D0i9QTQbuqps","execution_count":null,"outputs":[]}],"metadata":{"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"},"colab":{"provenance":[]}},"nbformat":4,"nbformat_minor":5}